# Sharp-wave ripples and hippocampal replay in DANDI:000044

This notebook demonstrates the two halves of the sharp-wave ripple (SWR) phenomenon
in real recordings from the DANDI Archive:

1. **The oscillation.** SWRs are 150-200 Hz transients in the CA1 pyramidal layer,
   riding on a slow sharp-wave deflection, that occur during slow-wave sleep and
   quiet immobility and never during theta states (running, REM).
2. **The content.** The population spikes packed inside a ripple are not random: they
   re-express the sequence of place cells that fired while the animal ran the track,
   compressed roughly ten-fold in time. This is *replay*, and it appears in the sleep
   that follows the run but not in the sleep that precedes it.

**Dataset.** [DANDI:000044](https://dandiarchive.org/dandiset/000044), Grosmark &
Buzsáki (2016), *Science* 351:1440. Four rats ran back and forth on a linear track
between two sleep sessions in the home cage. Each NWB file contains ~9 h of 128-channel
CA1 LFP at 1250 Hz, spike-sorted units labelled excitatory/inhibitory, the linearized
position on the track, and manually scored brain states (Awake / non-REM / REM).

**Access.** Files are ~9 GB each and are never downloaded whole. They are streamed with
`remfile` over a local disk cache; because the LFP dataset is chunked one channel per
chunk, reading a single channel for the whole session costs about 90 MB.

**Tooling.** All time-series handling (interval sets, restriction, tuning curves,
perievent alignment, wavelet transform, Bayesian decoding) uses
[Pynapple](https://pynapple.org). The analysis functions live in `pipeline.py` and the
streaming loaders in `swr_utils.py`, both in this directory.

In [1]:
import os
import warnings

import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pynapple as nap
import requests
from scipy import stats
from scipy.signal import welch

import pipeline as pl
import swr_utils as su

matplotlib.use("Agg")  # headless: figures are written to disk, never shown
warnings.filterwarnings("ignore", category=FutureWarning)

SESSION = "Achilles-10252013"
RNG = np.random.default_rng(1)

## 1. The dandiset

Eight sessions from four rats. We prototype on `Achilles-10252013` (the largest unit
yield) and repeat the full analysis on three more sessions at the end.

In [2]:
assets = requests.get(
    "https://api.dandiarchive.org/api/dandisets/000044/versions/draft/assets/",
    params={"page_size": 100},
).json()["results"]
for a in sorted(assets, key=lambda a: a["path"]):
    print("%-58s %5.1f GB" % (a["path"], a["size"] / 1e9))

sub-Achilles/sub-Achilles_ses-Achilles-10252013_behavior+ecephys.nwb   8.7 GB
sub-Achilles/sub-Achilles_ses-Achilles-11012013_behavior+ecephys.nwb   9.2 GB
sub-Buddy/sub-Buddy_ses-Buddy-06272013_behavior+ecephys.nwb   5.2 GB
sub-Cicero/sub-Cicero_ses-Cicero-09012014_behavior+ecephys.nwb   8.7 GB
sub-Cicero/sub-Cicero_ses-Cicero-09102014_behavior+ecephys.nwb   9.1 GB
sub-Cicero/sub-Cicero_ses-Cicero-09172014_behavior+ecephys.nwb   8.4 GB
sub-Gatsby/sub-Gatsby_ses-Gatsby-08022013_behavior+ecephys.nwb   7.9 GB
sub-Gatsby/sub-Gatsby_ses-Gatsby-08282013_behavior+ecephys.nwb   8.4 GB


## 2. Loading and inspecting the data streams

Every stream is checked before anything is computed from it.

In [3]:
h5 = su.open_session(SESSION)
epochs = su.load_epochs(h5)
states = su.load_states(h5)
units = su.load_units(h5)
pos = su.load_position(h5)
lfp_ds, fs, t0 = su.lfp_meta(h5)

pyr = units.getby_category("cell_type")["excitatory"]
inh = units.getby_category("cell_type")["inhibitory"]

print("session:", SESSION)
print("epochs:", {k: "%.0f-%.0f s" % (v.start[0], v.end[-1]) for k, v in epochs.items()})
print("states:", {k: "%d bouts, %.0f s" % (len(v), v.tot_length()) for k, v in states.items()})
print("units: %d excitatory, %d inhibitory (all CA1)" % (len(pyr), len(inh)))
print("LFP: %d samples x %d channels at %g Hz = %.1f h"
      % (lfp_ds.shape[0], lfp_ds.shape[1], fs, lfp_ds.shape[0] / fs / 3600))
print("position: %d samples, %.0f-%.0f cm (linearized only during traversals: %.0f%% NaN)"
      % (len(pos), np.nanmin(pos.values), np.nanmax(pos.values),
         100 * np.mean(~np.isfinite(pos.values))))

session: Achilles-10252013
epochs: {'PRE': '0-18080 s', 'MAZE': '18080-20147 s', 'POST': '20147-34861 s'}
states: {np.str_('Awake'): '62 bouts, 16747 s', np.str_('Non-REM'): '60 bouts, 15890 s', np.str_('REM'): '22 bouts, 2071 s'}
units: 120 excitatory, 17 inhibitory (all CA1)
LFP: 43576379 samples x 128 channels at 1250 Hz = 9.7 h
position: 80762 samples, 0-160 cm (linearized only during traversals: 87% NaN)


### Choosing a ripple channel

Ripple amplitude depends steeply on depth: it is maximal in the CA1 pyramidal layer and
falls off within a few hundred microns. The electrode table in this dandiset gives no
layer labels, so the channel is chosen empirically as the one with the largest
130-250 Hz envelope during a 200 s bout of post-task non-REM sleep. The profile below
shows the expected sawtooth: each of the 14 shanks has one site near the cell layer.

In [4]:
best_ch, alt_ch, sd, group = pl.pick_ripple_channel(h5, epochs, states)
print("ripple channel %d (%s); independent channel %d (%s)"
      % (best_ch, group[best_ch], alt_ch, group[alt_ch]))

fig, ax = plt.subplots(figsize=(11, 3.2))
ax.bar(np.arange(len(sd)), sd, color=["C0" if g == group[best_ch] else "0.6" for g in group])
for ch, c in ((best_ch, "C3"), (alt_ch, "C1")):
    ax.plot(ch, sd[ch] * 1.06, "v", color=c, ms=9)
    ax.text(ch, sd[ch] * 1.12, "ch %d" % ch, color=c, ha="center")
ax.margins(y=0.18)
ax.set(xlabel="LFP channel", ylabel="130-250 Hz envelope SD (a.u.)",
       title="%s: ripple-band power across 128 sites (200 s of non-REM sleep)" % SESSION)
fig.tight_layout()
fig.savefig("fig01_channel_selection.png", dpi=150)
plt.close(fig)

ripple channel 2 (shank1); independent channel 10 (shank2)


### Raw streams

Four seconds of sleep LFP, the same trace band-passed, the animal's position on the
track, and a spike raster. The band-passed trace already shows discrete high-frequency
bursts, which is what the detector below formalizes.

In [5]:
nrem_post = states["Non-REM"].intersect(epochs["POST"])
seg = int(np.argmax(nrem_post.end - nrem_post.start))
t_ex = float(nrem_post.start[seg]) + 25
lfp_ex = su.load_lfp_channel(h5, best_ch, t_ex, t_ex + 4)
filt_ex = pl.bandpass(lfp_ex.values, pl.LOW, pl.HIGH, fs)

fig, axes = plt.subplots(4, 1, figsize=(11, 9))
axes[0].plot(lfp_ex.index - t_ex, lfp_ex.values, lw=0.6, color="k")
axes[0].set(ylabel="LFP (a.u.)", title="Raw CA1 LFP, ch %d (non-REM sleep)" % best_ch)
axes[1].plot(lfp_ex.index - t_ex, filt_ex, lw=0.6, color="C3")
axes[1].set(ylabel="130-250 Hz", xlabel="time in window (s)")

maze = epochs["MAZE"]
p = pos.restrict(maze)
axes[2].plot(p.index, p.values, lw=0.8, color="C0")
axes[2].set(ylabel="position (cm)", xlabel="time (s)",
            title="Linearized position on the 1.6 m track (MAZE epoch)")

sub = pyr.restrict(nap.IntervalSet(maze.start[0], maze.start[0] + 120))
for i, k in enumerate(list(sub.keys())[:60]):
    tt = sub[k].index
    axes[3].plot(tt, np.full_like(tt, i), "|", ms=2.5, color="k", mew=0.5)
axes[3].set(ylabel="unit #", xlabel="time (s)",
            title="Spike raster, 60 pyramidal cells (first 2 min on the track)")
fig.tight_layout()
fig.savefig("fig02_raw_streams.png", dpi=150)
plt.close(fig)

## 3. Detecting sharp-wave ripples

Standard envelope-threshold detection on the whole 9.7 h recording: band-pass
130-250 Hz, Hilbert envelope, z-score, keep excursions whose peak exceeds 4 SD with
boundaries at 2 SD, merge events closer than 20 ms, keep durations of 20-200 ms.
Events during locomotion (speed > 4 cm/s) are discarded, since running-speed
high-frequency power is not a ripple.

In [6]:
ripples, peak_t, peak_z, raw, filt, fs, t0 = pl.ripples_from_channel(h5, best_ch, pos)
dur_ms = (ripples.end - ripples.start) * 1000
peak_freq = pl.ripple_stats(filt, fs, t0, peak_t)

print("%d ripples over %.1f h (%.2f Hz overall)"
      % (len(peak_t), (t0 + len(raw) / fs) / 3600, len(peak_t) / (len(raw) / fs)))
print("duration %.1f +- %.1f ms, intra-ripple frequency %.1f +- %.1f Hz"
      % (dur_ms.mean(), dur_ms.std(), peak_freq.mean(), peak_freq.std()))

rates = {}
for name, ep in list(epochs.items()) + list(states.items()):
    n = int(((peak_t[:, None] >= ep.start) & (peak_t[:, None] <= ep.end)).any(1).sum())
    rates[name] = n / ep.tot_length()
    print("  %-8s n=%5d  %8.0f s  %.3f Hz" % (name, n, ep.tot_length(), rates[name]))

14625 ripples over 9.7 h (0.42 Hz overall)
duration 48.6 +- 26.1 ms, intra-ripple frequency 164.7 +- 20.5 Hz
  PRE      n= 7405     18080 s  0.410 Hz
  MAZE     n=  451      2068 s  0.218 Hz
  POST     n= 6769     14714 s  0.460 Hz
  Awake    n= 4832     16747 s  0.289 Hz
  Non-REM  n= 9735     15890 s  0.613 Hz
  REM      n=    3      2071 s  0.001 Hz


### Are these really ripples?

Six independent checks, none of which the detector was tuned to satisfy:

- the event-triggered average of the **raw** LFP shows the slow sharp wave the ripple rides on;
- the mean wavelet spectrum has a single blob centred at ~160 Hz and ~50 ms wide;
- durations and intra-ripple frequencies match the textbook rodent values (~50 ms, 150-200 Hz);
- both pyramidal cells and interneurons increase their firing sharply and symmetrically
  around the ripple peak, interneurons much more strongly;
- the rate is high in non-REM, intermediate in quiet waking, and essentially zero in REM,
  which is the classic state-dependence and cannot come from an amplitude artifact;
- detecting independently on a channel from a different shank recovers the same events.

In [7]:
HALF = 0.25
n_half = int(HALF * fs)
idx = np.round((peak_t - t0) * fs).astype(int)
idx = idx[(idx > n_half) & (idx < len(raw) - n_half - 1)]
offs = np.arange(-n_half, n_half + 1)
lag = offs / fs
snips_raw = raw[idx[:, None] + offs]
snips_raw = snips_raw - snips_raw[:, :int(0.05 * fs)].mean(axis=1, keepdims=True)
snips_filt = filt[idx[:, None] + offs]

freqs = np.geomspace(30, 400, 60)
sub_ev = RNG.choice(len(snips_raw), size=min(800, len(snips_raw)), replace=False)
tf = np.zeros((len(freqs), snips_raw.shape[1]))
for i in sub_ev:
    sig = nap.Tsd(t=lag - lag[0] + 1.0, d=snips_raw[i].astype(float))
    tf += np.abs(np.asarray(nap.compute_wavelet_transform(sig, freqs, fs=fs))).T ** 2
tf /= len(sub_ev)

peaks = nap.Ts(t=peak_t)
psth = {}
for name, grp in (("pyramidal", pyr), ("interneuron", inh)):
    pe = nap.compute_perievent(grp, peaks, window=(-HALF, HALF))
    r = []
    for k in pe.keys():
        c = pe[k].count(0.005, nap.IntervalSet(-HALF, HALF))
        r.append(np.asarray(c).sum(axis=1) / (0.005 * len(peak_t)))
    psth[name] = (np.asarray(c.index), np.array(r))

# same detection on a channel from a different shank
lfp2 = su.load_lfp_channel(h5, alt_ch)
filt2 = pl.bandpass(lfp2.values, pl.LOW, pl.HIGH, fs)
env2 = pl.envelope(filt2, fs)
_, peak_t2, _ = pl.detect_ripples(nap.Tsd(t=lfp2.index, d=(env2 - env2.mean()) / env2.std()))
nearest = peak_t2[np.argmin(np.abs(peak_t2[None, :] - peak_t[:, None]), axis=1)] - peak_t
match = np.abs(nearest) < 0.05
print("channel %d: %d ripples, %.1f%% of channel-%d events matched within 50 ms"
      % (alt_ch, len(peak_t2), 100 * match.mean(), best_ch))

channel 10: 14384 ripples, 88.8% of channel-2 events matched within 50 ms


In [8]:
fig = plt.figure(figsize=(13, 9))
gs = fig.add_gridspec(3, 3, hspace=0.55, wspace=0.32)

ax = fig.add_subplot(gs[0, 0])
m, s = snips_raw.mean(0), snips_raw.std(0) / np.sqrt(len(snips_raw))
ax.plot(lag * 1000, m, color="k", lw=1.2)
ax.fill_between(lag * 1000, m - s, m + s, color="k", alpha=0.3)
ax.axvline(0, color="C3", ls=":", lw=1)
ax.set(xlabel="time from ripple peak (ms)", ylabel="LFP (a.u.)",
       title="Ripple-triggered average LFP\n(the sharp wave)")

ax = fig.add_subplot(gs[0, 1])
ax.plot(lag * 1000, snips_filt.mean(0), color="C3", lw=1)
ax.set(xlabel="time from ripple peak (ms)", ylabel="130-250 Hz (a.u.)",
       title="Ripple-triggered average\nof the filtered band", xlim=(-100, 100))

ax = fig.add_subplot(gs[0, 2])
im = ax.pcolormesh(lag * 1000, freqs, tf, shading="auto", cmap="magma")
ax.axhline(130, color="w", ls=":", lw=0.8)
ax.axhline(250, color="w", ls=":", lw=0.8)
ax.set(yscale="log", xlabel="time from ripple peak (ms)", ylabel="frequency (Hz)",
       title="Mean wavelet power\n(n=%d events)" % len(sub_ev), xlim=(-150, 150))
fig.colorbar(im, ax=ax, label="power (a.u.)")

ax = fig.add_subplot(gs[1, 0])
ax.hist(dur_ms, bins=40, color="C0")
ax.set(xlabel="duration (ms)", ylabel="count",
       title="Event duration\nmedian %.0f ms" % np.median(dur_ms))

ax = fig.add_subplot(gs[1, 1])
ax.hist(peak_freq, bins=40, color="C0")
ax.set(xlabel="intra-ripple peak frequency (Hz)", ylabel="count",
       title="Ripple frequency\nmedian %.0f Hz" % np.median(peak_freq))

ax = fig.add_subplot(gs[1, 2])
ax.hist(peak_z, bins=np.arange(4, 25, 0.5), color="C0")
ax.set(xlabel="peak envelope (z)", ylabel="count", title="Ripple amplitude", yscale="log")

ax = fig.add_subplot(gs[2, 0])
for name, color in (("pyramidal", "C0"), ("interneuron", "C1")):
    tt, r = psth[name]
    mm, se = r.mean(0), r.std(0) / np.sqrt(len(r))
    ax.plot(tt * 1000, mm, color=color, label="%s (n=%d)" % (name, len(r)))
    ax.fill_between(tt * 1000, mm - se, mm + se, color=color, alpha=0.3)
ax.axvline(0, color="0.5", ls=":")
ax.legend(fontsize=8, frameon=False)
ax.set(xlabel="time from ripple peak (ms)", ylabel="firing rate (Hz)",
       title="Spiking is locked to ripples")

ax = fig.add_subplot(gs[2, 1])
names = list(epochs) + list(states)
ax.bar(names, [rates[n] for n in names], color=["C0"] * 3 + ["C2"] * 3)
ax.set_xticks(range(len(names)))
ax.set_xticklabels(names, rotation=35, ha="right")
ax.set(ylabel="ripple rate (Hz)", title="Ripples are a non-REM /\nimmobility phenomenon")

ax = fig.add_subplot(gs[2, 2])
ax.hist(nearest, bins=np.arange(-0.05, 0.0501, 0.004), color="C0")
ax.set(xlabel="peak-time difference (s)", ylabel="count",
       title="Same events on the shank of ch %d\n(%.0f%% matched)" % (alt_ch, 100 * match.mean()))

fig.suptitle("Sharp-wave ripples in CA1, %s (DANDI:000044), n=%d events" % (SESSION, len(peak_t)))
fig.savefig("fig03_ripple_characterization.png", dpi=150, bbox_inches="tight")
plt.close(fig)

### Individual events

In [9]:
order = np.argsort(peak_z)[::-1]
in_nrem = ((peak_t[:, None] >= nrem_post.start) & (peak_t[:, None] <= nrem_post.end)).any(1)
sel = [i for i in order if in_nrem[i]][:6]

fig, axes = plt.subplots(2, 3, figsize=(13, 6.5), sharex=True)
for ax, i in zip(axes.ravel(), sel):
    tc = peak_t[i]
    sl = slice(int((tc - 0.2 - t0) * fs), int((tc + 0.2 - t0) * fs))
    tt = (np.arange(sl.start, sl.stop) / fs + t0 - tc) * 1000
    ax.plot(tt, raw[sl] - raw[sl].mean(), color="k", lw=0.7)
    ax.plot(tt, filt[sl] - 2200, color="C3", lw=0.7)
    sp = pyr.restrict(nap.IntervalSet(tc - 0.2, tc + 0.2))
    for j, k in enumerate(sp.keys()):
        s = (sp[k].index - tc) * 1000
        ax.plot(s, np.full_like(s, -3200 - 22 * j), "|", color="C0", ms=3, mew=0.7)
    ax.axvspan((ripples.start[i] - tc) * 1000, (ripples.end[i] - tc) * 1000, color="C3", alpha=0.12)
    ax.set(title="peak %.1f z, t=%.1f s" % (peak_z[i], tc), yticks=[])
    ax.set_xlabel("time from ripple peak (ms)")
for ax in axes[:, 0]:
    ax.set_ylabel("LFP / filtered / spikes")
fig.suptitle("Example sharp-wave ripples during post-task non-REM sleep (raster: pyramidal cells)")
fig.tight_layout()
fig.savefig("fig04_example_ripples.png", dpi=150)
plt.close(fig)

## 4. Place fields: the template that replay is scored against

Replay can only be read out against a map of what the cells code for. Rate maps are
built from running periods only (speed > 5 cm/s) and separately for the two travel
directions, because CA1 fields on a linear track are strongly direction-selective.

In [10]:
pos_maze = pos[np.isfinite(pos.values)].restrict(epochs["MAZE"])
track_len = float(np.ceil(np.nanmax(pos.values)))
run, right, left = pl.running_epochs(pos_maze)
tcs, centers, si, is_place = pl.place_fields(pyr, pos_maze, right, left, track_len)
ids = np.array(list(pyr.keys()))[is_place]
spk = pyr[list(ids)]
templates = {k: pl.make_template(tcs[k][:, is_place], ids, centers) for k in tcs}

print("%d rightward runs (%.0f s), %d leftward runs (%.0f s)"
      % (len(right), right.tot_length(), len(left), left.tot_length()))
print("%d of %d pyramidal cells pass the place-cell criteria; median spatial information %.2f bits/spike"
      % (is_place.sum(), len(pyr), np.median(np.maximum(si["right"], si["left"])[is_place])))

56 rightward runs (159 s), 47 leftward runs (132 s)
106 of 120 pyramidal cells pass the place-cell criteria; median spatial information 1.20 bits/spike


In [11]:
fig = plt.figure(figsize=(13, 7.5))
gs = fig.add_gridspec(2, 3, hspace=0.45, wspace=0.35, height_ratios=[1.25, 1])
idx_pc = np.where(is_place)[0]

for j, k in enumerate(["right", "left"]):
    ax = fig.add_subplot(gs[0, j])
    o = idx_pc[np.argsort(np.argmax(tcs[k][:, idx_pc], axis=0))]
    m = tcs[k][:, o].T
    im = ax.imshow(m / m.max(1, keepdims=True), aspect="auto", origin="lower", cmap="viridis",
                   extent=[0, track_len, 0, len(o)])
    ax.set(xlabel="position on track (cm)", ylabel="place cell (sorted by own peak)",
           title="%sward runs" % k.capitalize())
    fig.colorbar(im, ax=ax, label="normalized rate")

ax = fig.add_subplot(gs[0, 2])
pk_all = centers[np.argmax(tcs["right"], axis=0)]
mid = idx_pc[(pk_all[idx_pc] > 20) & (pk_all[idx_pc] < track_len - 20)]
for c in mid[np.argsort(-si["right"][mid])][:6]:
    ax.plot(centers, tcs["right"][:, c], lw=1.4)
ax.set(xlabel="position (cm)", ylabel="firing rate (Hz)",
       title="Six example place fields\n(rightward runs)")

ax = fig.add_subplot(gs[1, 0])
p = pos_maze.restrict(nap.IntervalSet(pos_maze.index[0], pos_maze.index[0] + 300))
ax.plot(p.index - p.index[0], p.values, color="0.5", lw=0.8)
for ep_dir, c in ((right, "C0"), (left, "C3")):
    for s, e in zip(ep_dir.start, ep_dir.end):
        sg = pos_maze.restrict(nap.IntervalSet(s, e))
        ax.plot(sg.index - p.index[0], sg.values, color=c, lw=1.6)
ax.set(xlim=(0, 300), xlabel="time on maze (s)", ylabel="position (cm)",
       title="Laps: rightward (blue) / leftward (red)")

ax = fig.add_subplot(gs[1, 1])
si_max = np.maximum(si["right"], si["left"])
ax.hist(si_max, bins=30, color="0.7", label="all pyramidal")
ax.hist(si_max[is_place], bins=30, color="C0", label="place cells")
ax.axvline(pl.MIN_SPATIAL_INFO, color="k", ls=":")
ax.legend(fontsize=8, frameon=False)
ax.set(xlabel="spatial information (bits/spike)", ylabel="count", title="Spatial tuning")

ax = fig.add_subplot(gs[1, 2])
pk_r = centers[np.argmax(tcs["right"][:, is_place], axis=0)]
pk_l = centers[np.argmax(tcs["left"][:, is_place], axis=0)]
ax.plot(pk_r, pk_l, "o", ms=4, alpha=0.6)
ax.plot([0, track_len], [0, track_len], "k:", lw=1)
ax.set(xlabel="peak position, rightward (cm)", ylabel="peak position, leftward (cm)",
       title="Fields are direction-specific\n(r=%.2f)" % np.corrcoef(pk_r, pk_l)[0, 1])

fig.suptitle("Place fields on the linear track, %s" % SESSION)
fig.savefig("fig05_place_fields.png", dpi=150, bbox_inches="tight")
plt.close(fig)

### Does the template actually decode position?

Before asking what the posterior says during sleep, we check what it says during
running, where ground truth exists. A median error of a few centimetres on a 160 cm
track (chance is ~47 cm) means the template is sound.

In [12]:
errs = []
for k, ep_dir in (("right", right), ("left", left)):
    dec, _ = nap.decode_bayes(tuning_curves=templates[k], data=spk, epochs=ep_dir, bin_size=0.25)
    errs.append(np.abs(np.asarray(dec) - np.asarray(pos_maze.interpolate(dec))))
err = np.concatenate(errs)
chance = np.median(np.abs(RNG.uniform(0, track_len, 20000) - RNG.uniform(0, track_len, 20000)))
run_speed = float(np.median(su.compute_speed(pos_maze).restrict(run).values))
print("decoder median error while running: %.1f cm (chance %.1f cm); median running speed %.0f cm/s"
      % (np.median(err), chance, run_speed))

decoder median error while running: 4.7 cm (chance 46.8 cm); median running speed 42 cm/s


## 5. Replay inside ripples

**Candidate events.** Population bursts of the place-cell ensemble (smoothed multi-unit
rate above its mean, peak > 3 SD, 100-500 ms long) that contain a detected ripple peak
and in which at least 5 place cells fire. Requiring both the LFP signature and the
population burst is what ties the sequence content to the ripple.

**Scoring.** Each event is decoded in 20 ms bins with a uniform-prior Bayesian decoder
and scored by the weighted correlation between decoded position and time. Both
direction templates are tried and the larger |r| is kept.

**Significance.** Two shuffles of the posterior, 400 draws each: a per-bin circular
shift in position (destroys trajectory continuity, keeps per-bin structure) and a
permutation of time bins (destroys temporal order). Because the observed statistic is a
maximum over two templates, the null is the same maximum over the two templates'
matched shuffles, so nothing is gained for free by trying both. An event counts as
replay when it beats both nulls at p < 0.05.

That last step is nominally a 5% false-positive rate, but it should not be trusted as
one. Posterior shuffles preserve some of the structure that produces apparent sequences,
so the achieved rate is measured empirically with the cell-identity shuffle below rather
than assumed. It comes out at 6 to 9% depending on session and random draw, so every
comparison in this notebook is made against that empirical null and not against 0.05.

In [13]:
rows, posteriors = [], {}
for ep_name in ("PRE", "POST"):
    pbe = pl.population_bursts(spk, epochs[ep_name], peak_t)
    n_rip = int(((peak_t >= epochs[ep_name].start[0]) & (peak_t <= epochs[ep_name].end[-1])).sum())
    print("%s: %d ripple-associated population bursts (of %d ripples)" % (ep_name, len(pbe), n_rip))
    r, p_ = pl.score_epoch(spk, templates, centers, pbe, ep_name, RNG)
    rows += r
    posteriors.update(p_)

df = pd.DataFrame(rows)
df["significant"] = df["max_p"] < pl.ALPHA
df["fwd"] = ((df.template == "right") & (df.slope > 0)) | ((df.template == "left") & (df.slope < 0))
df.to_csv("replay_events.csv", index=False)

PRE: 3932 ripple-associated population bursts (of 7405 ripples)


decoding PRE:   0%|          | 0/3932 [00:00<?, ?it/s]

decoding PRE:   1%|          | 31/3932 [00:00<00:12, 308.23it/s]

decoding PRE:   2%|▏         | 62/3932 [00:00<00:28, 135.19it/s]

decoding PRE:   2%|▏         | 81/3932 [00:00<00:30, 127.41it/s]

decoding PRE:   2%|▏         | 97/3932 [00:00<00:30, 125.65it/s]

decoding PRE:   3%|▎         | 117/3932 [00:00<00:26, 141.84it/s]

decoding PRE:   3%|▎         | 133/3932 [00:00<00:28, 135.34it/s]

decoding PRE:   4%|▍         | 148/3932 [00:01<00:31, 120.12it/s]

decoding PRE:   4%|▍         | 165/3932 [00:01<00:30, 122.40it/s]

decoding PRE:   5%|▍         | 180/3932 [00:01<00:29, 128.98it/s]

decoding PRE:   5%|▌         | 197/3932 [00:01<00:26, 138.76it/s]

decoding PRE:   5%|▌         | 212/3932 [00:01<00:30, 120.13it/s]

decoding PRE:   6%|▌         | 225/3932 [00:01<00:32, 115.08it/s]

decoding PRE:   6%|▌         | 238/3932 [00:01<00:33, 111.37it/s]

decoding PRE:   7%|▋         | 257/3932 [00:01<00:28, 128.41it/s]

decoding PRE:   7%|▋         | 271/3932 [00:02<00:29, 123.11it/s]

decoding PRE:   7%|▋         | 284/3932 [00:02<00:30, 118.46it/s]

decoding PRE:   8%|▊         | 297/3932 [00:02<00:36, 99.09it/s] 

decoding PRE:   8%|▊         | 323/3932 [00:02<00:28, 128.70it/s]

decoding PRE:   9%|▊         | 340/3932 [00:02<00:26, 138.02it/s]

decoding PRE:   9%|▉         | 357/3932 [00:02<00:26, 136.18it/s]

decoding PRE:   9%|▉         | 372/3932 [00:02<00:29, 120.68it/s]

decoding PRE:  10%|▉         | 385/3932 [00:03<00:30, 115.10it/s]

decoding PRE:  10%|█         | 397/3932 [00:03<00:32, 110.09it/s]

decoding PRE:  11%|█         | 415/3932 [00:03<00:27, 126.86it/s]

decoding PRE:  11%|█         | 429/3932 [00:03<00:33, 106.15it/s]

decoding PRE:  11%|█▏        | 445/3932 [00:03<00:31, 111.73it/s]

decoding PRE:  12%|█▏        | 457/3932 [00:03<00:32, 108.45it/s]

decoding PRE:  12%|█▏        | 472/3932 [00:03<00:31, 110.12it/s]

decoding PRE:  12%|█▏        | 484/3932 [00:03<00:32, 106.57it/s]

decoding PRE:  13%|█▎        | 501/3932 [00:04<00:28, 121.80it/s]

decoding PRE:  13%|█▎        | 514/3932 [00:04<00:29, 115.36it/s]

decoding PRE:  13%|█▎        | 526/3932 [00:04<00:35, 96.25it/s] 

decoding PRE:  14%|█▍        | 541/3932 [00:04<00:33, 102.70it/s]

decoding PRE:  14%|█▍        | 558/3932 [00:04<00:28, 118.26it/s]

decoding PRE:  15%|█▍        | 575/3932 [00:04<00:25, 131.05it/s]

decoding PRE:  15%|█▌        | 591/3932 [00:04<00:24, 137.87it/s]

decoding PRE:  15%|█▌        | 609/3932 [00:04<00:22, 148.45it/s]

decoding PRE:  16%|█▌        | 625/3932 [00:05<00:25, 128.03it/s]

decoding PRE:  16%|█▋        | 639/3932 [00:05<00:27, 121.95it/s]

decoding PRE:  17%|█▋        | 652/3932 [00:05<00:28, 116.15it/s]

decoding PRE:  17%|█▋        | 668/3932 [00:05<00:27, 117.18it/s]

decoding PRE:  17%|█▋        | 683/3932 [00:05<00:25, 125.28it/s]

decoding PRE:  18%|█▊        | 696/3932 [00:05<00:27, 119.00it/s]

decoding PRE:  18%|█▊        | 709/3932 [00:05<00:33, 96.91it/s] 

decoding PRE:  18%|█▊        | 727/3932 [00:06<00:27, 115.21it/s]

decoding PRE:  19%|█▉        | 740/3932 [00:06<00:33, 96.53it/s] 

decoding PRE:  19%|█▉        | 756/3932 [00:06<00:28, 110.33it/s]

decoding PRE:  20%|█▉        | 771/3932 [00:06<00:26, 119.56it/s]

decoding PRE:  20%|█▉        | 785/3932 [00:06<00:27, 114.60it/s]

decoding PRE:  20%|██        | 798/3932 [00:06<00:28, 111.86it/s]

decoding PRE:  21%|██        | 810/3932 [00:06<00:29, 106.72it/s]

decoding PRE:  21%|██        | 822/3932 [00:06<00:30, 103.23it/s]

decoding PRE:  21%|██        | 833/3932 [00:07<00:33, 91.93it/s] 

decoding PRE:  21%|██▏       | 844/3932 [00:07<00:33, 91.62it/s]

decoding PRE:  22%|██▏       | 858/3932 [00:07<00:31, 96.86it/s]

decoding PRE:  22%|██▏       | 868/3932 [00:07<00:37, 81.04it/s]

decoding PRE:  22%|██▏       | 882/3932 [00:07<00:34, 87.83it/s]

decoding PRE:  23%|██▎       | 893/3932 [00:07<00:34, 89.10it/s]

decoding PRE:  23%|██▎       | 903/3932 [00:07<00:35, 85.07it/s]

decoding PRE:  23%|██▎       | 918/3932 [00:07<00:29, 100.50it/s]

decoding PRE:  24%|██▎       | 929/3932 [00:08<00:34, 87.89it/s] 

decoding PRE:  24%|██▍       | 950/3932 [00:08<00:26, 113.71it/s]

decoding PRE:  25%|██▍       | 966/3932 [00:08<00:25, 116.64it/s]

decoding PRE:  25%|██▍       | 979/3932 [00:08<00:31, 93.34it/s] 

decoding PRE:  26%|██▌       | 1009/3932 [00:08<00:21, 138.03it/s]

decoding PRE:  26%|██▌       | 1026/3932 [00:08<00:24, 117.43it/s]

decoding PRE:  26%|██▋       | 1040/3932 [00:09<00:26, 107.15it/s]

decoding PRE:  27%|██▋       | 1053/3932 [00:09<00:32, 89.05it/s] 

decoding PRE:  27%|██▋       | 1064/3932 [00:09<00:38, 75.35it/s]

decoding PRE:  27%|██▋       | 1079/3932 [00:09<00:32, 88.81it/s]

decoding PRE:  28%|██▊       | 1090/3932 [00:09<00:33, 83.98it/s]

decoding PRE:  28%|██▊       | 1100/3932 [00:09<00:34, 83.03it/s]

decoding PRE:  28%|██▊       | 1118/3932 [00:10<00:27, 103.63it/s]

decoding PRE:  29%|██▊       | 1130/3932 [00:10<00:27, 100.78it/s]

decoding PRE:  29%|██▉       | 1142/3932 [00:10<00:27, 100.29it/s]

decoding PRE:  29%|██▉       | 1153/3932 [00:10<00:37, 73.85it/s] 

decoding PRE:  30%|██▉       | 1163/3932 [00:10<00:36, 76.22it/s]

decoding PRE:  30%|██▉       | 1177/3932 [00:10<00:32, 84.40it/s]

decoding PRE:  30%|███       | 1187/3932 [00:10<00:31, 86.87it/s]

decoding PRE:  30%|███       | 1197/3932 [00:11<00:37, 73.81it/s]

decoding PRE:  31%|███       | 1206/3932 [00:11<00:36, 74.29it/s]

decoding PRE:  31%|███       | 1220/3932 [00:11<00:32, 84.46it/s]

decoding PRE:  31%|███▏      | 1229/3932 [00:11<00:32, 82.58it/s]

decoding PRE:  31%|███▏      | 1238/3932 [00:11<00:35, 74.93it/s]

decoding PRE:  32%|███▏      | 1246/3932 [00:11<00:36, 73.71it/s]

decoding PRE:  32%|███▏      | 1254/3932 [00:11<00:36, 73.20it/s]

decoding PRE:  32%|███▏      | 1262/3932 [00:11<00:40, 66.42it/s]

decoding PRE:  32%|███▏      | 1270/3932 [00:12<00:40, 66.22it/s]

decoding PRE:  33%|███▎      | 1287/3932 [00:12<00:28, 91.72it/s]

decoding PRE:  33%|███▎      | 1297/3932 [00:12<00:31, 82.48it/s]

decoding PRE:  33%|███▎      | 1308/3932 [00:12<00:30, 85.07it/s]

decoding PRE:  34%|███▎      | 1319/3932 [00:12<00:30, 86.03it/s]

decoding PRE:  34%|███▍      | 1329/3932 [00:12<00:30, 85.13it/s]

decoding PRE:  34%|███▍      | 1338/3932 [00:12<00:37, 70.02it/s]

decoding PRE:  34%|███▍      | 1346/3932 [00:12<00:36, 70.37it/s]

decoding PRE:  35%|███▍      | 1359/3932 [00:13<00:32, 78.71it/s]

decoding PRE:  35%|███▍      | 1368/3932 [00:13<00:35, 71.68it/s]

decoding PRE:  35%|███▍      | 1376/3932 [00:13<00:36, 70.69it/s]

decoding PRE:  35%|███▌      | 1389/3932 [00:13<00:31, 80.17it/s]

decoding PRE:  36%|███▌      | 1398/3932 [00:13<00:32, 77.94it/s]

decoding PRE:  36%|███▌      | 1406/3932 [00:13<00:33, 75.72it/s]

decoding PRE:  36%|███▌      | 1416/3932 [00:13<00:32, 78.20it/s]

decoding PRE:  36%|███▌      | 1424/3932 [00:13<00:32, 76.36it/s]

decoding PRE:  36%|███▋      | 1432/3932 [00:14<00:36, 69.07it/s]

decoding PRE:  37%|███▋      | 1448/3932 [00:14<00:27, 91.92it/s]

decoding PRE:  37%|███▋      | 1459/3932 [00:14<00:26, 91.90it/s]

decoding PRE:  37%|███▋      | 1469/3932 [00:14<00:27, 89.50it/s]

decoding PRE:  38%|███▊      | 1481/3932 [00:14<00:26, 91.84it/s]

decoding PRE:  38%|███▊      | 1496/3932 [00:14<00:22, 106.99it/s]

decoding PRE:  38%|███▊      | 1508/3932 [00:14<00:26, 90.46it/s] 

decoding PRE:  39%|███▊      | 1518/3932 [00:14<00:27, 88.54it/s]

decoding PRE:  39%|███▉      | 1533/3932 [00:15<00:23, 103.17it/s]

decoding PRE:  39%|███▉      | 1544/3932 [00:15<00:26, 91.05it/s] 

decoding PRE:  40%|███▉      | 1554/3932 [00:15<00:26, 88.84it/s]

decoding PRE:  40%|███▉      | 1569/3932 [00:15<00:22, 103.52it/s]

decoding PRE:  40%|████      | 1580/3932 [00:15<00:25, 91.47it/s] 

decoding PRE:  40%|████      | 1590/3932 [00:15<00:26, 89.03it/s]

decoding PRE:  41%|████      | 1605/3932 [00:15<00:22, 103.84it/s]

decoding PRE:  41%|████      | 1616/3932 [00:15<00:23, 98.70it/s] 

decoding PRE:  42%|████▏     | 1636/3932 [00:16<00:18, 121.26it/s]

decoding PRE:  42%|████▏     | 1649/3932 [00:16<00:25, 87.96it/s] 

decoding PRE:  42%|████▏     | 1663/3932 [00:16<00:24, 92.78it/s]

decoding PRE:  43%|████▎     | 1674/3932 [00:16<00:24, 92.16it/s]

decoding PRE:  43%|████▎     | 1685/3932 [00:16<00:24, 90.61it/s]

decoding PRE:  43%|████▎     | 1696/3932 [00:16<00:24, 89.96it/s]

decoding PRE:  43%|████▎     | 1706/3932 [00:16<00:25, 87.08it/s]

decoding PRE:  44%|████▎     | 1717/3932 [00:17<00:25, 87.20it/s]

decoding PRE:  44%|████▍     | 1726/3932 [00:17<00:28, 77.63it/s]

decoding PRE:  44%|████▍     | 1734/3932 [00:17<00:29, 74.57it/s]

decoding PRE:  44%|████▍     | 1742/3932 [00:17<00:32, 67.78it/s]

decoding PRE:  44%|████▍     | 1749/3932 [00:17<00:36, 60.29it/s]

decoding PRE:  45%|████▍     | 1760/3932 [00:17<00:32, 67.84it/s]

decoding PRE:  45%|████▍     | 1767/3932 [00:18<00:40, 53.95it/s]

decoding PRE:  45%|████▌     | 1777/3932 [00:18<00:33, 63.42it/s]

decoding PRE:  45%|████▌     | 1785/3932 [00:18<00:32, 66.78it/s]

decoding PRE:  46%|████▌     | 1795/3932 [00:18<00:28, 74.20it/s]

decoding PRE:  46%|████▌     | 1809/3932 [00:18<00:25, 84.47it/s]

decoding PRE:  46%|████▌     | 1818/3932 [00:18<00:25, 81.83it/s]

decoding PRE:  46%|████▋     | 1827/3932 [00:18<00:25, 81.08it/s]

decoding PRE:  47%|████▋     | 1838/3932 [00:18<00:24, 83.86it/s]

decoding PRE:  47%|████▋     | 1855/3932 [00:18<00:20, 103.84it/s]

decoding PRE:  47%|████▋     | 1867/3932 [00:19<00:20, 101.27it/s]

decoding PRE:  48%|████▊     | 1878/3932 [00:19<00:21, 97.24it/s] 

decoding PRE:  48%|████▊     | 1888/3932 [00:19<00:27, 73.58it/s]

decoding PRE:  48%|████▊     | 1897/3932 [00:19<00:27, 73.78it/s]

decoding PRE:  48%|████▊     | 1905/3932 [00:19<00:30, 67.14it/s]

decoding PRE:  49%|████▊     | 1913/3932 [00:19<00:29, 67.63it/s]

decoding PRE:  49%|████▉     | 1921/3932 [00:19<00:29, 68.07it/s]

decoding PRE:  49%|████▉     | 1935/3932 [00:20<00:23, 85.66it/s]

decoding PRE:  50%|████▉     | 1949/3932 [00:20<00:21, 92.00it/s]

decoding PRE:  50%|████▉     | 1959/3932 [00:20<00:22, 89.15it/s]

decoding PRE:  50%|█████     | 1969/3932 [00:20<00:22, 87.32it/s]

decoding PRE:  50%|█████     | 1978/3932 [00:20<00:23, 82.69it/s]

decoding PRE:  51%|█████     | 1993/3932 [00:20<00:19, 99.71it/s]

decoding PRE:  51%|█████     | 2004/3932 [00:20<00:19, 96.40it/s]

decoding PRE:  51%|█████▏    | 2024/3932 [00:20<00:15, 119.96it/s]

decoding PRE:  52%|█████▏    | 2037/3932 [00:20<00:16, 113.58it/s]

decoding PRE:  52%|█████▏    | 2049/3932 [00:21<00:17, 107.08it/s]

decoding PRE:  52%|█████▏    | 2060/3932 [00:21<00:18, 101.28it/s]

decoding PRE:  53%|█████▎    | 2071/3932 [00:21<00:19, 97.68it/s] 

decoding PRE:  53%|█████▎    | 2081/3932 [00:21<00:21, 87.06it/s]

decoding PRE:  53%|█████▎    | 2090/3932 [00:21<00:26, 68.61it/s]

decoding PRE:  54%|█████▎    | 2104/3932 [00:21<00:23, 78.23it/s]

decoding PRE:  54%|█████▎    | 2113/3932 [00:21<00:24, 73.10it/s]

decoding PRE:  54%|█████▍    | 2126/3932 [00:22<00:22, 80.21it/s]

decoding PRE:  54%|█████▍    | 2138/3932 [00:22<00:21, 84.24it/s]

decoding PRE:  55%|█████▍    | 2151/3932 [00:22<00:19, 89.36it/s]

decoding PRE:  55%|█████▍    | 2162/3932 [00:22<00:19, 90.05it/s]

decoding PRE:  55%|█████▌    | 2175/3932 [00:22<00:19, 92.12it/s]

decoding PRE:  56%|█████▌    | 2189/3932 [00:22<00:18, 96.17it/s]

decoding PRE:  56%|█████▌    | 2202/3932 [00:22<00:17, 97.72it/s]

decoding PRE:  56%|█████▋    | 2216/3932 [00:23<00:15, 107.74it/s]

decoding PRE:  57%|█████▋    | 2232/3932 [00:23<00:14, 120.63it/s]

decoding PRE:  57%|█████▋    | 2245/3932 [00:23<00:15, 111.86it/s]

decoding PRE:  58%|█████▊    | 2261/3932 [00:23<00:13, 123.75it/s]

decoding PRE:  58%|█████▊    | 2274/3932 [00:23<00:15, 107.83it/s]

decoding PRE:  58%|█████▊    | 2286/3932 [00:23<00:16, 97.76it/s] 

decoding PRE:  58%|█████▊    | 2297/3932 [00:23<00:17, 95.23it/s]

decoding PRE:  59%|█████▊    | 2307/3932 [00:23<00:17, 91.40it/s]

decoding PRE:  59%|█████▉    | 2317/3932 [00:24<00:18, 88.21it/s]

decoding PRE:  59%|█████▉    | 2326/3932 [00:24<00:21, 73.10it/s]

decoding PRE:  59%|█████▉    | 2334/3932 [00:24<00:25, 62.36it/s]

decoding PRE:  60%|█████▉    | 2351/3932 [00:24<00:18, 84.16it/s]

decoding PRE:  60%|██████    | 2361/3932 [00:24<00:18, 84.45it/s]

decoding PRE:  60%|██████    | 2371/3932 [00:24<00:21, 72.48it/s]

decoding PRE:  61%|██████    | 2380/3932 [00:24<00:21, 73.32it/s]

decoding PRE:  61%|██████    | 2390/3932 [00:25<00:20, 76.06it/s]

decoding PRE:  61%|██████    | 2406/3932 [00:25<00:15, 95.95it/s]

decoding PRE:  62%|██████▏   | 2420/3932 [00:25<00:15, 98.59it/s]

decoding PRE:  62%|██████▏   | 2431/3932 [00:25<00:19, 76.88it/s]

decoding PRE:  62%|██████▏   | 2442/3932 [00:25<00:18, 79.74it/s]

decoding PRE:  62%|██████▏   | 2455/3932 [00:25<00:17, 84.33it/s]

decoding PRE:  63%|██████▎   | 2465/3932 [00:25<00:20, 73.26it/s]

decoding PRE:  63%|██████▎   | 2473/3932 [00:26<00:20, 72.07it/s]

decoding PRE:  63%|██████▎   | 2481/3932 [00:26<00:24, 59.18it/s]

decoding PRE:  63%|██████▎   | 2490/3932 [00:26<00:22, 62.71it/s]

decoding PRE:  64%|██████▎   | 2502/3932 [00:26<00:20, 71.39it/s]

decoding PRE:  64%|██████▍   | 2510/3932 [00:26<00:21, 64.67it/s]

decoding PRE:  64%|██████▍   | 2518/3932 [00:26<00:21, 65.05it/s]

decoding PRE:  64%|██████▍   | 2531/3932 [00:26<00:18, 75.58it/s]

decoding PRE:  65%|██████▍   | 2543/3932 [00:27<00:17, 80.05it/s]

decoding PRE:  65%|██████▍   | 2554/3932 [00:27<00:16, 81.90it/s]

decoding PRE:  65%|██████▌   | 2563/3932 [00:27<00:17, 79.91it/s]

decoding PRE:  65%|██████▌   | 2572/3932 [00:27<00:17, 77.58it/s]

decoding PRE:  66%|██████▌   | 2583/3932 [00:27<00:16, 80.91it/s]

decoding PRE:  66%|██████▌   | 2593/3932 [00:27<00:16, 79.97it/s]

decoding PRE:  66%|██████▌   | 2604/3932 [00:27<00:16, 82.04it/s]

decoding PRE:  66%|██████▋   | 2613/3932 [00:27<00:17, 73.45it/s]

decoding PRE:  67%|██████▋   | 2621/3932 [00:28<00:18, 71.36it/s]

decoding PRE:  67%|██████▋   | 2629/3932 [00:28<00:20, 65.08it/s]

decoding PRE:  67%|██████▋   | 2636/3932 [00:28<00:21, 59.06it/s]

decoding PRE:  67%|██████▋   | 2643/3932 [00:28<00:24, 53.13it/s]

decoding PRE:  68%|██████▊   | 2655/3932 [00:28<00:19, 64.37it/s]

decoding PRE:  68%|██████▊   | 2663/3932 [00:28<00:19, 65.42it/s]

decoding PRE:  68%|██████▊   | 2676/3932 [00:28<00:16, 75.34it/s]

decoding PRE:  68%|██████▊   | 2686/3932 [00:29<00:16, 76.95it/s]

decoding PRE:  69%|██████▊   | 2701/3932 [00:29<00:12, 94.84it/s]

decoding PRE:  69%|██████▉   | 2711/3932 [00:29<00:13, 91.44it/s]

decoding PRE:  69%|██████▉   | 2725/3932 [00:29<00:12, 95.33it/s]

decoding PRE:  70%|██████▉   | 2737/3932 [00:29<00:12, 96.13it/s]

decoding PRE:  70%|██████▉   | 2747/3932 [00:29<00:12, 91.59it/s]

decoding PRE:  70%|███████   | 2762/3932 [00:29<00:11, 97.64it/s]

decoding PRE:  70%|███████   | 2772/3932 [00:29<00:14, 80.15it/s]

decoding PRE:  71%|███████   | 2781/3932 [00:30<00:14, 78.22it/s]

decoding PRE:  71%|███████   | 2795/3932 [00:30<00:13, 86.39it/s]

decoding PRE:  71%|███████▏  | 2804/3932 [00:30<00:13, 83.78it/s]

decoding PRE:  72%|███████▏  | 2817/3932 [00:30<00:12, 89.09it/s]

decoding PRE:  72%|███████▏  | 2837/3932 [00:30<00:09, 112.52it/s]

decoding PRE:  73%|███████▎  | 2852/3932 [00:30<00:08, 121.90it/s]

decoding PRE:  73%|███████▎  | 2865/3932 [00:30<00:09, 114.60it/s]

decoding PRE:  73%|███████▎  | 2877/3932 [00:31<00:10, 99.51it/s] 

decoding PRE:  74%|███████▎  | 2892/3932 [00:31<00:09, 110.92it/s]

decoding PRE:  74%|███████▍  | 2904/3932 [00:31<00:09, 105.53it/s]

decoding PRE:  74%|███████▍  | 2916/3932 [00:31<00:09, 109.12it/s]

decoding PRE:  75%|███████▍  | 2931/3932 [00:31<00:08, 119.53it/s]

decoding PRE:  75%|███████▍  | 2944/3932 [00:31<00:10, 96.41it/s] 

decoding PRE:  75%|███████▌  | 2955/3932 [00:31<00:11, 82.15it/s]

decoding PRE:  75%|███████▌  | 2967/3932 [00:31<00:11, 85.80it/s]

decoding PRE:  76%|███████▌  | 2981/3932 [00:32<00:09, 97.80it/s]

decoding PRE:  76%|███████▋  | 2999/3932 [00:32<00:08, 114.85it/s]

decoding PRE:  77%|███████▋  | 3012/3932 [00:32<00:07, 118.26it/s]

decoding PRE:  77%|███████▋  | 3025/3932 [00:32<00:08, 110.01it/s]

decoding PRE:  77%|███████▋  | 3037/3932 [00:32<00:08, 104.64it/s]

decoding PRE:  78%|███████▊  | 3048/3932 [00:32<00:10, 85.75it/s] 

decoding PRE:  78%|███████▊  | 3062/3932 [00:32<00:08, 97.66it/s]

decoding PRE:  78%|███████▊  | 3073/3932 [00:32<00:09, 95.26it/s]

decoding PRE:  79%|███████▊  | 3089/3932 [00:33<00:07, 110.29it/s]

decoding PRE:  79%|███████▉  | 3109/3932 [00:33<00:06, 128.77it/s]

decoding PRE:  80%|███████▉  | 3126/3932 [00:33<00:05, 137.08it/s]

decoding PRE:  80%|███████▉  | 3141/3932 [00:33<00:07, 110.98it/s]

decoding PRE:  80%|████████  | 3154/3932 [00:33<00:08, 93.73it/s] 

decoding PRE:  81%|████████  | 3168/3932 [00:33<00:07, 103.34it/s]

decoding PRE:  81%|████████  | 3180/3932 [00:33<00:07, 100.26it/s]

decoding PRE:  81%|████████  | 3192/3932 [00:34<00:07, 99.33it/s] 

decoding PRE:  82%|████████▏ | 3206/3932 [00:34<00:06, 109.18it/s]

decoding PRE:  82%|████████▏ | 3221/3932 [00:34<00:05, 119.43it/s]

decoding PRE:  82%|████████▏ | 3238/3932 [00:34<00:05, 130.99it/s]

decoding PRE:  83%|████████▎ | 3252/3932 [00:34<00:06, 105.19it/s]

decoding PRE:  83%|████████▎ | 3264/3932 [00:34<00:07, 88.75it/s] 

decoding PRE:  83%|████████▎ | 3274/3932 [00:34<00:07, 86.65it/s]

decoding PRE:  84%|████████▎ | 3284/3932 [00:35<00:08, 74.21it/s]

decoding PRE:  84%|████████▎ | 3293/3932 [00:35<00:09, 65.82it/s]

decoding PRE:  84%|████████▍ | 3311/3932 [00:35<00:07, 87.09it/s]

decoding PRE:  85%|████████▍ | 3326/3932 [00:35<00:06, 100.63it/s]

decoding PRE:  85%|████████▍ | 3338/3932 [00:35<00:06, 91.92it/s] 

decoding PRE:  85%|████████▌ | 3351/3932 [00:35<00:06, 94.71it/s]

decoding PRE:  86%|████████▌ | 3362/3932 [00:35<00:05, 97.89it/s]

decoding PRE:  86%|████████▌ | 3373/3932 [00:36<00:06, 81.55it/s]

decoding PRE:  86%|████████▌ | 3382/3932 [00:36<00:07, 68.91it/s]

decoding PRE:  86%|████████▋ | 3396/3932 [00:36<00:06, 78.18it/s]

decoding PRE:  87%|████████▋ | 3406/3932 [00:36<00:06, 78.35it/s]

decoding PRE:  87%|████████▋ | 3422/3932 [00:36<00:05, 95.89it/s]

decoding PRE:  87%|████████▋ | 3433/3932 [00:36<00:06, 81.88it/s]

decoding PRE:  88%|████████▊ | 3449/3932 [00:36<00:04, 97.53it/s]

decoding PRE:  88%|████████▊ | 3463/3932 [00:36<00:04, 107.46it/s]

decoding PRE:  88%|████████▊ | 3475/3932 [00:37<00:04, 102.83it/s]

decoding PRE:  89%|████████▉ | 3495/3932 [00:37<00:03, 123.37it/s]

decoding PRE:  89%|████████▉ | 3509/3932 [00:37<00:03, 108.61it/s]

decoding PRE:  90%|████████▉ | 3523/3932 [00:37<00:03, 107.49it/s]

decoding PRE:  90%|████████▉ | 3535/3932 [00:37<00:04, 89.99it/s] 

decoding PRE:  90%|█████████ | 3546/3932 [00:37<00:04, 88.87it/s]

decoding PRE:  91%|█████████ | 3560/3932 [00:37<00:04, 92.70it/s]

decoding PRE:  91%|█████████ | 3572/3932 [00:38<00:03, 93.03it/s]

decoding PRE:  91%|█████████ | 3584/3932 [00:38<00:03, 93.42it/s]

decoding PRE:  91%|█████████▏| 3597/3932 [00:38<00:03, 101.61it/s]

decoding PRE:  92%|█████████▏| 3608/3932 [00:38<00:03, 89.26it/s] 

decoding PRE:  92%|█████████▏| 3618/3932 [00:38<00:04, 77.02it/s]

decoding PRE:  92%|█████████▏| 3634/3932 [00:38<00:03, 90.93it/s]

decoding PRE:  93%|█████████▎| 3644/3932 [00:39<00:03, 76.63it/s]

decoding PRE:  93%|█████████▎| 3656/3932 [00:39<00:03, 79.79it/s]

decoding PRE:  93%|█████████▎| 3665/3932 [00:39<00:04, 65.53it/s]

decoding PRE:  93%|█████████▎| 3673/3932 [00:39<00:03, 65.28it/s]

decoding PRE:  94%|█████████▎| 3680/3932 [00:39<00:04, 59.13it/s]

decoding PRE:  94%|█████████▍| 3691/3932 [00:39<00:03, 65.99it/s]

decoding PRE:  94%|█████████▍| 3698/3932 [00:39<00:03, 63.79it/s]

decoding PRE:  94%|█████████▍| 3711/3932 [00:40<00:03, 73.28it/s]

decoding PRE:  95%|█████████▍| 3719/3932 [00:40<00:03, 62.61it/s]

decoding PRE:  95%|█████████▍| 3726/3932 [00:40<00:03, 61.85it/s]

decoding PRE:  95%|█████████▌| 3739/3932 [00:40<00:02, 72.48it/s]

decoding PRE:  96%|█████████▌| 3758/3932 [00:40<00:01, 97.79it/s]

decoding PRE:  96%|█████████▌| 3771/3932 [00:40<00:01, 98.27it/s]

decoding PRE:  96%|█████████▌| 3782/3932 [00:40<00:01, 94.51it/s]

decoding PRE:  97%|█████████▋| 3798/3932 [00:40<00:01, 109.20it/s]

decoding PRE:  97%|█████████▋| 3810/3932 [00:41<00:01, 104.26it/s]

decoding PRE:  97%|█████████▋| 3821/3932 [00:41<00:01, 75.25it/s] 

decoding PRE:  97%|█████████▋| 3830/3932 [00:41<00:01, 56.23it/s]

decoding PRE:  98%|█████████▊| 3839/3932 [00:41<00:01, 59.57it/s]

decoding PRE:  98%|█████████▊| 3847/3932 [00:41<00:01, 60.75it/s]

decoding PRE:  98%|█████████▊| 3854/3932 [00:42<00:01, 49.86it/s]

decoding PRE:  98%|█████████▊| 3860/3932 [00:42<00:01, 50.11it/s]

decoding PRE:  98%|█████████▊| 3868/3932 [00:42<00:01, 54.42it/s]

decoding PRE:  99%|█████████▊| 3876/3932 [00:42<00:00, 57.71it/s]

decoding PRE:  99%|█████████▉| 3883/3932 [00:42<00:00, 58.38it/s]

decoding PRE:  99%|█████████▉| 3897/3932 [00:42<00:00, 78.26it/s]

decoding PRE:  99%|█████████▉| 3906/3932 [00:42<00:00, 70.51it/s]

decoding PRE: 100%|█████████▉| 3916/3932 [00:42<00:00, 73.29it/s]

decoding PRE: 100%|█████████▉| 3924/3932 [00:43<00:00, 71.80it/s]

decoding PRE: 100%|██████████| 3932/3932 [00:43<00:00, 91.15it/s]

POST: 3345 ripple-associated population bursts (of 6769 ripples)


decoding POST:   0%|          | 0/3345 [00:00<?, ?it/s]

decoding POST:   0%|          | 8/3345 [00:00<00:51, 64.41it/s]

decoding POST:   0%|          | 16/3345 [00:00<00:50, 65.53it/s]

decoding POST:   1%|          | 30/3345 [00:00<00:35, 94.23it/s]

decoding POST:   1%|          | 40/3345 [00:00<00:42, 77.86it/s]

decoding POST:   2%|▏         | 53/3345 [00:00<00:38, 84.50it/s]

decoding POST:   2%|▏         | 64/3345 [00:00<00:38, 85.89it/s]

decoding POST:   2%|▏         | 73/3345 [00:00<00:39, 82.01it/s]

decoding POST:   2%|▏         | 83/3345 [00:01<00:39, 81.56it/s]

decoding POST:   3%|▎         | 100/3345 [00:01<00:32, 101.09it/s]

decoding POST:   3%|▎         | 114/3345 [00:01<00:29, 111.02it/s]

decoding POST:   4%|▍         | 126/3345 [00:01<00:30, 103.87it/s]

decoding POST:   4%|▍         | 137/3345 [00:01<00:32, 98.74it/s] 

decoding POST:   4%|▍         | 149/3345 [00:01<00:33, 96.41it/s]

decoding POST:   5%|▍         | 162/3345 [00:01<00:32, 97.15it/s]

decoding POST:   5%|▌         | 172/3345 [00:01<00:34, 92.06it/s]

decoding POST:   6%|▌         | 186/3345 [00:02<00:32, 95.84it/s]

decoding POST:   6%|▌         | 196/3345 [00:02<00:34, 90.03it/s]

decoding POST:   6%|▋         | 212/3345 [00:02<00:29, 104.62it/s]

decoding POST:   7%|▋         | 223/3345 [00:02<00:31, 99.30it/s] 

decoding POST:   7%|▋         | 239/3345 [00:02<00:27, 112.25it/s]

decoding POST:   8%|▊         | 252/3345 [00:02<00:26, 115.21it/s]

decoding POST:   8%|▊         | 264/3345 [00:02<00:28, 108.09it/s]

decoding POST:   8%|▊         | 275/3345 [00:02<00:30, 99.66it/s] 

decoding POST:   9%|▊         | 288/3345 [00:02<00:28, 107.44it/s]

decoding POST:   9%|▉         | 300/3345 [00:03<00:30, 101.23it/s]

decoding POST:   9%|▉         | 311/3345 [00:03<00:38, 79.60it/s] 

decoding POST:  10%|▉         | 320/3345 [00:03<00:41, 73.30it/s]

decoding POST:  10%|█         | 337/3345 [00:03<00:32, 92.46it/s]

decoding POST:  11%|█         | 358/3345 [00:03<00:26, 114.30it/s]

decoding POST:  11%|█         | 372/3345 [00:03<00:26, 111.26it/s]

decoding POST:  12%|█▏        | 385/3345 [00:03<00:25, 115.49it/s]

decoding POST:  12%|█▏        | 398/3345 [00:04<00:35, 83.33it/s] 

decoding POST:  12%|█▏        | 408/3345 [00:04<00:39, 75.27it/s]

decoding POST:  12%|█▏        | 417/3345 [00:04<00:39, 74.90it/s]

decoding POST:  13%|█▎        | 426/3345 [00:04<00:41, 70.33it/s]

decoding POST:  13%|█▎        | 440/3345 [00:04<00:34, 85.37it/s]

decoding POST:  14%|█▎        | 452/3345 [00:04<00:30, 93.56it/s]

decoding POST:  14%|█▍        | 463/3345 [00:04<00:31, 90.76it/s]

decoding POST:  14%|█▍        | 477/3345 [00:05<00:27, 102.77it/s]

decoding POST:  15%|█▍        | 488/3345 [00:05<00:29, 97.34it/s] 

decoding POST:  15%|█▍        | 500/3345 [00:05<00:29, 96.30it/s]

decoding POST:  15%|█▌        | 511/3345 [00:05<00:30, 92.66it/s]

decoding POST:  16%|█▌        | 521/3345 [00:05<00:34, 81.55it/s]

decoding POST:  16%|█▌        | 530/3345 [00:05<00:35, 79.49it/s]

decoding POST:  16%|█▌        | 539/3345 [00:05<00:36, 77.76it/s]

decoding POST:  16%|█▋        | 549/3345 [00:06<00:35, 78.00it/s]

decoding POST:  17%|█▋        | 557/3345 [00:06<00:37, 74.50it/s]

decoding POST:  17%|█▋        | 569/3345 [00:06<00:32, 85.87it/s]

decoding POST:  17%|█▋        | 579/3345 [00:06<00:33, 83.66it/s]

decoding POST:  18%|█▊        | 589/3345 [00:06<00:33, 81.80it/s]

decoding POST:  18%|█▊        | 598/3345 [00:06<00:37, 72.40it/s]

decoding POST:  18%|█▊        | 606/3345 [00:06<00:41, 66.08it/s]

decoding POST:  18%|█▊        | 617/3345 [00:06<00:37, 71.83it/s]

decoding POST:  19%|█▊        | 625/3345 [00:07<00:38, 69.79it/s]

decoding POST:  19%|█▉        | 633/3345 [00:07<00:42, 63.93it/s]

decoding POST:  19%|█▉        | 646/3345 [00:07<00:33, 79.52it/s]

decoding POST:  20%|█▉        | 658/3345 [00:07<00:32, 83.59it/s]

decoding POST:  20%|█▉        | 667/3345 [00:07<00:38, 69.25it/s]

decoding POST:  20%|██        | 679/3345 [00:07<00:33, 80.49it/s]

decoding POST:  21%|██        | 691/3345 [00:07<00:31, 83.25it/s]

decoding POST:  21%|██        | 700/3345 [00:08<00:38, 68.56it/s]

decoding POST:  21%|██        | 708/3345 [00:08<00:41, 63.73it/s]

decoding POST:  21%|██▏       | 715/3345 [00:08<00:44, 58.87it/s]

decoding POST:  22%|██▏       | 734/3345 [00:08<00:30, 84.31it/s]

decoding POST:  22%|██▏       | 744/3345 [00:08<00:33, 76.93it/s]

decoding POST:  23%|██▎       | 753/3345 [00:08<00:39, 66.37it/s]

decoding POST:  23%|██▎       | 767/3345 [00:08<00:31, 81.65it/s]

decoding POST:  23%|██▎       | 780/3345 [00:09<00:27, 92.68it/s]

decoding POST:  24%|██▍       | 796/3345 [00:09<00:23, 107.47it/s]

decoding POST:  24%|██▍       | 808/3345 [00:09<00:26, 94.31it/s] 

decoding POST:  24%|██▍       | 819/3345 [00:09<00:29, 85.25it/s]

decoding POST:  25%|██▍       | 832/3345 [00:09<00:26, 95.24it/s]

decoding POST:  25%|██▌       | 843/3345 [00:09<00:29, 85.66it/s]

decoding POST:  26%|██▌       | 855/3345 [00:09<00:26, 93.46it/s]

decoding POST:  26%|██▌       | 866/3345 [00:09<00:27, 91.06it/s]

decoding POST:  26%|██▌       | 876/3345 [00:10<00:30, 81.22it/s]

decoding POST:  26%|██▋       | 885/3345 [00:10<00:33, 73.26it/s]

decoding POST:  27%|██▋       | 898/3345 [00:10<00:28, 86.20it/s]

decoding POST:  27%|██▋       | 908/3345 [00:10<00:33, 72.90it/s]

decoding POST:  27%|██▋       | 917/3345 [00:10<00:33, 72.88it/s]

decoding POST:  28%|██▊       | 929/3345 [00:10<00:28, 83.58it/s]

decoding POST:  28%|██▊       | 941/3345 [00:10<00:28, 85.41it/s]

decoding POST:  29%|██▊       | 954/3345 [00:11<00:24, 96.21it/s]

decoding POST:  29%|██▉       | 967/3345 [00:11<00:22, 104.95it/s]

decoding POST:  29%|██▉       | 979/3345 [00:11<00:27, 86.33it/s] 

decoding POST:  30%|██▉       | 991/3345 [00:11<00:25, 94.13it/s]

decoding POST:  30%|██▉       | 1002/3345 [00:11<00:29, 79.57it/s]

decoding POST:  30%|███       | 1011/3345 [00:11<00:29, 78.03it/s]

decoding POST:  31%|███       | 1024/3345 [00:11<00:27, 82.91it/s]

decoding POST:  31%|███       | 1033/3345 [00:12<00:29, 78.89it/s]

decoding POST:  31%|███       | 1042/3345 [00:12<00:31, 72.30it/s]

decoding POST:  32%|███▏      | 1054/3345 [00:12<00:29, 77.78it/s]

decoding POST:  32%|███▏      | 1067/3345 [00:12<00:25, 89.49it/s]

decoding POST:  32%|███▏      | 1077/3345 [00:12<00:29, 78.07it/s]

decoding POST:  33%|███▎      | 1088/3345 [00:12<00:26, 85.48it/s]

decoding POST:  33%|███▎      | 1098/3345 [00:12<00:27, 82.92it/s]

decoding POST:  33%|███▎      | 1107/3345 [00:12<00:28, 79.37it/s]

decoding POST:  34%|███▎      | 1122/3345 [00:13<00:23, 93.70it/s]

decoding POST:  34%|███▍      | 1132/3345 [00:13<00:25, 87.99it/s]

decoding POST:  34%|███▍      | 1142/3345 [00:13<00:26, 84.42it/s]

decoding POST:  34%|███▍      | 1151/3345 [00:13<00:33, 65.07it/s]

decoding POST:  35%|███▍      | 1159/3345 [00:13<00:33, 64.84it/s]

decoding POST:  35%|███▍      | 1170/3345 [00:13<00:31, 69.88it/s]

decoding POST:  35%|███▌      | 1179/3345 [00:13<00:29, 74.45it/s]

decoding POST:  35%|███▌      | 1187/3345 [00:14<00:30, 71.45it/s]

decoding POST:  36%|███▌      | 1198/3345 [00:14<00:29, 74.01it/s]

decoding POST:  36%|███▋      | 1216/3345 [00:14<00:22, 96.32it/s]

decoding POST:  37%|███▋      | 1226/3345 [00:14<00:25, 84.45it/s]

decoding POST:  37%|███▋      | 1237/3345 [00:14<00:23, 90.43it/s]

decoding POST:  37%|███▋      | 1247/3345 [00:14<00:24, 87.41it/s]

decoding POST:  38%|███▊      | 1258/3345 [00:14<00:24, 86.25it/s]

decoding POST:  38%|███▊      | 1270/3345 [00:14<00:21, 94.70it/s]

decoding POST:  38%|███▊      | 1280/3345 [00:15<00:24, 82.76it/s]

decoding POST:  39%|███▊      | 1289/3345 [00:15<00:31, 64.98it/s]

decoding POST:  39%|███▉      | 1300/3345 [00:15<00:27, 74.36it/s]

decoding POST:  39%|███▉      | 1311/3345 [00:15<00:26, 76.47it/s]

decoding POST:  39%|███▉      | 1320/3345 [00:15<00:26, 76.10it/s]

decoding POST:  40%|███▉      | 1329/3345 [00:15<00:30, 65.60it/s]

decoding POST:  40%|███▉      | 1337/3345 [00:15<00:30, 65.87it/s]

decoding POST:  40%|████      | 1344/3345 [00:16<00:31, 63.16it/s]

decoding POST:  40%|████      | 1351/3345 [00:16<00:32, 62.03it/s]

decoding POST:  41%|████      | 1364/3345 [00:16<00:27, 72.94it/s]

decoding POST:  41%|████▏     | 1388/3345 [00:16<00:18, 105.51it/s]

decoding POST:  42%|████▏     | 1401/3345 [00:16<00:17, 110.77it/s]

decoding POST:  42%|████▏     | 1413/3345 [00:16<00:21, 88.43it/s] 

decoding POST:  43%|████▎     | 1424/3345 [00:16<00:21, 87.86it/s]

decoding POST:  43%|████▎     | 1434/3345 [00:17<00:22, 84.59it/s]

decoding POST:  43%|████▎     | 1449/3345 [00:17<00:19, 97.89it/s]

decoding POST:  44%|████▎     | 1460/3345 [00:17<00:20, 93.54it/s]

decoding POST:  44%|████▍     | 1473/3345 [00:17<00:19, 94.77it/s]

decoding POST:  44%|████▍     | 1483/3345 [00:17<00:20, 90.90it/s]

decoding POST:  45%|████▍     | 1496/3345 [00:17<00:18, 100.17it/s]

decoding POST:  45%|████▌     | 1507/3345 [00:17<00:20, 88.16it/s] 

decoding POST:  46%|████▌     | 1523/3345 [00:17<00:17, 102.34it/s]

decoding POST:  46%|████▌     | 1534/3345 [00:18<00:20, 88.82it/s] 

decoding POST:  46%|████▌     | 1544/3345 [00:18<00:25, 70.85it/s]

decoding POST:  46%|████▋     | 1552/3345 [00:18<00:26, 68.95it/s]

decoding POST:  47%|████▋     | 1568/3345 [00:18<00:20, 87.17it/s]

decoding POST:  47%|████▋     | 1580/3345 [00:18<00:18, 94.69it/s]

decoding POST:  48%|████▊     | 1591/3345 [00:18<00:23, 73.49it/s]

decoding POST:  48%|████▊     | 1600/3345 [00:18<00:23, 73.30it/s]

decoding POST:  48%|████▊     | 1609/3345 [00:19<00:23, 73.22it/s]

decoding POST:  48%|████▊     | 1620/3345 [00:19<00:22, 76.71it/s]

decoding POST:  49%|████▉     | 1633/3345 [00:19<00:20, 83.06it/s]

decoding POST:  49%|████▉     | 1647/3345 [00:19<00:17, 94.59it/s]

decoding POST:  50%|████▉     | 1666/3345 [00:19<00:14, 112.44it/s]

decoding POST:  50%|█████     | 1678/3345 [00:19<00:15, 106.16it/s]

decoding POST:  50%|█████     | 1689/3345 [00:19<00:16, 98.71it/s] 

decoding POST:  51%|█████     | 1700/3345 [00:20<00:18, 88.36it/s]

decoding POST:  51%|█████     | 1711/3345 [00:20<00:17, 93.41it/s]

decoding POST:  52%|█████▏    | 1725/3345 [00:20<00:15, 103.98it/s]

decoding POST:  52%|█████▏    | 1737/3345 [00:20<00:16, 99.88it/s] 

decoding POST:  52%|█████▏    | 1748/3345 [00:20<00:19, 82.00it/s]

decoding POST:  53%|█████▎    | 1773/3345 [00:20<00:14, 111.32it/s]

decoding POST:  53%|█████▎    | 1785/3345 [00:20<00:17, 91.41it/s] 

decoding POST:  54%|█████▎    | 1797/3345 [00:20<00:15, 96.82it/s]

decoding POST:  54%|█████▍    | 1808/3345 [00:21<00:16, 93.65it/s]

decoding POST:  54%|█████▍    | 1820/3345 [00:21<00:16, 92.27it/s]

decoding POST:  55%|█████▍    | 1830/3345 [00:21<00:18, 82.26it/s]

decoding POST:  55%|█████▍    | 1839/3345 [00:21<00:18, 79.81it/s]

decoding POST:  55%|█████▌    | 1851/3345 [00:21<00:16, 89.02it/s]

decoding POST:  56%|█████▌    | 1861/3345 [00:21<00:18, 79.49it/s]

decoding POST:  56%|█████▌    | 1872/3345 [00:21<00:18, 80.97it/s]

decoding POST:  56%|█████▌    | 1881/3345 [00:22<00:18, 78.03it/s]

decoding POST:  57%|█████▋    | 1893/3345 [00:22<00:17, 82.35it/s]

decoding POST:  57%|█████▋    | 1902/3345 [00:22<00:18, 79.80it/s]

decoding POST:  57%|█████▋    | 1913/3345 [00:22<00:17, 81.17it/s]

decoding POST:  57%|█████▋    | 1922/3345 [00:22<00:17, 79.49it/s]

decoding POST:  58%|█████▊    | 1930/3345 [00:22<00:20, 70.56it/s]

decoding POST:  58%|█████▊    | 1938/3345 [00:22<00:20, 68.99it/s]

decoding POST:  58%|█████▊    | 1949/3345 [00:22<00:19, 73.06it/s]

decoding POST:  59%|█████▊    | 1964/3345 [00:23<00:15, 90.49it/s]

decoding POST:  59%|█████▉    | 1979/3345 [00:23<00:13, 103.09it/s]

decoding POST:  60%|█████▉    | 1991/3345 [00:23<00:13, 98.61it/s] 

decoding POST:  60%|█████▉    | 2002/3345 [00:23<00:14, 94.13it/s]

decoding POST:  60%|██████    | 2012/3345 [00:23<00:15, 87.99it/s]

decoding POST:  60%|██████    | 2021/3345 [00:23<00:15, 85.78it/s]

decoding POST:  61%|██████    | 2033/3345 [00:23<00:15, 87.13it/s]

decoding POST:  61%|██████    | 2042/3345 [00:24<00:19, 65.61it/s]

decoding POST:  61%|██████▏   | 2050/3345 [00:24<00:19, 65.06it/s]

decoding POST:  62%|██████▏   | 2067/3345 [00:24<00:14, 85.40it/s]

decoding POST:  62%|██████▏   | 2080/3345 [00:24<00:13, 95.65it/s]

decoding POST:  63%|██████▎   | 2091/3345 [00:24<00:15, 80.19it/s]

decoding POST:  63%|██████▎   | 2102/3345 [00:24<00:14, 86.64it/s]

decoding POST:  63%|██████▎   | 2117/3345 [00:24<00:12, 100.35it/s]

decoding POST:  64%|██████▎   | 2129/3345 [00:24<00:12, 97.07it/s] 

decoding POST:  64%|██████▍   | 2140/3345 [00:25<00:12, 98.66it/s]

decoding POST:  64%|██████▍   | 2155/3345 [00:25<00:11, 106.68it/s]

decoding POST:  65%|██████▍   | 2167/3345 [00:25<00:10, 109.64it/s]

decoding POST:  65%|██████▌   | 2180/3345 [00:25<00:10, 114.71it/s]

decoding POST:  66%|██████▌   | 2192/3345 [00:25<00:12, 89.98it/s] 

decoding POST:  66%|██████▌   | 2202/3345 [00:25<00:15, 75.14it/s]

decoding POST:  66%|██████▌   | 2211/3345 [00:25<00:17, 65.37it/s]

decoding POST:  66%|██████▋   | 2219/3345 [00:26<00:17, 64.49it/s]

decoding POST:  67%|██████▋   | 2233/3345 [00:26<00:13, 80.16it/s]

decoding POST:  67%|██████▋   | 2242/3345 [00:26<00:15, 72.95it/s]

decoding POST:  67%|██████▋   | 2251/3345 [00:26<00:15, 72.86it/s]

decoding POST:  68%|██████▊   | 2262/3345 [00:26<00:16, 67.32it/s]

decoding POST:  68%|██████▊   | 2270/3345 [00:26<00:15, 67.87it/s]

decoding POST:  69%|██████▊   | 2292/3345 [00:26<00:10, 96.31it/s]

decoding POST:  69%|██████▉   | 2303/3345 [00:27<00:11, 92.83it/s]

decoding POST:  69%|██████▉   | 2313/3345 [00:27<00:13, 77.24it/s]

decoding POST:  69%|██████▉   | 2322/3345 [00:27<00:13, 75.99it/s]

decoding POST:  70%|██████▉   | 2330/3345 [00:27<00:15, 66.91it/s]

decoding POST:  70%|██████▉   | 2340/3345 [00:27<00:14, 70.26it/s]

decoding POST:  70%|███████   | 2348/3345 [00:27<00:14, 68.68it/s]

decoding POST:  70%|███████   | 2356/3345 [00:27<00:14, 67.13it/s]

decoding POST:  71%|███████   | 2363/3345 [00:28<00:15, 65.08it/s]

decoding POST:  71%|███████   | 2375/3345 [00:28<00:13, 72.36it/s]

decoding POST:  71%|███████   | 2383/3345 [00:28<00:15, 60.99it/s]

decoding POST:  71%|███████▏  | 2390/3345 [00:28<00:15, 60.19it/s]

decoding POST:  72%|███████▏  | 2401/3345 [00:28<00:14, 66.72it/s]

decoding POST:  72%|███████▏  | 2408/3345 [00:28<00:14, 63.83it/s]

decoding POST:  72%|███████▏  | 2415/3345 [00:28<00:14, 62.04it/s]

decoding POST:  72%|███████▏  | 2424/3345 [00:28<00:14, 64.48it/s]

decoding POST:  73%|███████▎  | 2431/3345 [00:29<00:14, 63.13it/s]

decoding POST:  73%|███████▎  | 2443/3345 [00:29<00:12, 71.50it/s]

decoding POST:  73%|███████▎  | 2454/3345 [00:29<00:11, 80.89it/s]

decoding POST:  74%|███████▎  | 2463/3345 [00:29<00:11, 78.38it/s]

decoding POST:  74%|███████▍  | 2475/3345 [00:29<00:10, 82.34it/s]

decoding POST:  74%|███████▍  | 2484/3345 [00:29<00:14, 59.57it/s]

decoding POST:  75%|███████▍  | 2497/3345 [00:29<00:11, 73.12it/s]

decoding POST:  75%|███████▍  | 2506/3345 [00:30<00:12, 67.53it/s]

decoding POST:  75%|███████▌  | 2516/3345 [00:30<00:11, 70.76it/s]

decoding POST:  76%|███████▌  | 2527/3345 [00:30<00:10, 79.67it/s]

decoding POST:  76%|███████▌  | 2536/3345 [00:30<00:10, 76.96it/s]

decoding POST:  76%|███████▌  | 2545/3345 [00:30<00:10, 74.83it/s]

decoding POST:  76%|███████▋  | 2554/3345 [00:30<00:10, 73.89it/s]

decoding POST:  77%|███████▋  | 2562/3345 [00:30<00:11, 70.92it/s]

decoding POST:  77%|███████▋  | 2570/3345 [00:30<00:11, 68.50it/s]

decoding POST:  77%|███████▋  | 2577/3345 [00:31<00:11, 65.37it/s]

decoding POST:  77%|███████▋  | 2584/3345 [00:31<00:12, 62.67it/s]

decoding POST:  77%|███████▋  | 2591/3345 [00:31<00:12, 60.39it/s]

decoding POST:  78%|███████▊  | 2598/3345 [00:31<00:13, 55.50it/s]

decoding POST:  78%|███████▊  | 2607/3345 [00:31<00:12, 60.16it/s]

decoding POST:  78%|███████▊  | 2615/3345 [00:31<00:11, 61.80it/s]

decoding POST:  78%|███████▊  | 2625/3345 [00:31<00:10, 66.06it/s]

decoding POST:  79%|███████▉  | 2638/3345 [00:31<00:08, 81.43it/s]

decoding POST:  79%|███████▉  | 2647/3345 [00:32<00:08, 78.31it/s]

decoding POST:  79%|███████▉  | 2656/3345 [00:32<00:09, 75.74it/s]

decoding POST:  80%|███████▉  | 2664/3345 [00:32<00:09, 72.79it/s]

decoding POST:  80%|███████▉  | 2672/3345 [00:32<00:11, 59.83it/s]

decoding POST:  80%|████████  | 2680/3345 [00:32<00:10, 60.47it/s]

decoding POST:  80%|████████  | 2688/3345 [00:32<00:10, 61.36it/s]

decoding POST:  81%|████████  | 2695/3345 [00:32<00:11, 56.38it/s]

decoding POST:  81%|████████  | 2701/3345 [00:33<00:11, 55.34it/s]

decoding POST:  81%|████████  | 2710/3345 [00:33<00:10, 59.83it/s]

decoding POST:  81%|████████▏ | 2721/3345 [00:33<00:09, 66.86it/s]

decoding POST:  82%|████████▏ | 2734/3345 [00:33<00:07, 81.92it/s]

decoding POST:  82%|████████▏ | 2743/3345 [00:33<00:07, 78.74it/s]

decoding POST:  82%|████████▏ | 2752/3345 [00:33<00:08, 68.93it/s]

decoding POST:  83%|████████▎ | 2761/3345 [00:33<00:08, 70.15it/s]

decoding POST:  83%|████████▎ | 2771/3345 [00:33<00:07, 76.50it/s]

decoding POST:  83%|████████▎ | 2779/3345 [00:34<00:08, 66.36it/s]

decoding POST:  83%|████████▎ | 2789/3345 [00:34<00:07, 69.61it/s]

decoding POST:  84%|████████▎ | 2797/3345 [00:34<00:07, 68.50it/s]

decoding POST:  84%|████████▍ | 2809/3345 [00:34<00:06, 80.69it/s]

decoding POST:  84%|████████▍ | 2819/3345 [00:34<00:06, 78.68it/s]

decoding POST:  85%|████████▍ | 2828/3345 [00:34<00:07, 71.65it/s]

decoding POST:  85%|████████▍ | 2839/3345 [00:34<00:06, 75.12it/s]

decoding POST:  85%|████████▌ | 2847/3345 [00:35<00:07, 66.98it/s]

decoding POST:  85%|████████▌ | 2857/3345 [00:35<00:06, 69.99it/s]

decoding POST:  86%|████████▌ | 2865/3345 [00:35<00:06, 68.59it/s]

decoding POST:  86%|████████▌ | 2874/3345 [00:35<00:06, 68.40it/s]

decoding POST:  86%|████████▌ | 2881/3345 [00:35<00:07, 60.28it/s]

decoding POST:  86%|████████▋ | 2888/3345 [00:35<00:08, 52.04it/s]

decoding POST:  87%|████████▋ | 2899/3345 [00:35<00:07, 60.44it/s]

decoding POST:  87%|████████▋ | 2911/3345 [00:36<00:05, 73.56it/s]

decoding POST:  87%|████████▋ | 2924/3345 [00:36<00:04, 85.41it/s]

decoding POST:  88%|████████▊ | 2934/3345 [00:36<00:05, 76.73it/s]

decoding POST:  88%|████████▊ | 2943/3345 [00:36<00:06, 65.61it/s]

decoding POST:  88%|████████▊ | 2951/3345 [00:36<00:06, 65.44it/s]

decoding POST:  88%|████████▊ | 2958/3345 [00:36<00:06, 59.28it/s]

decoding POST:  89%|████████▊ | 2965/3345 [00:36<00:07, 51.50it/s]

decoding POST:  89%|████████▉ | 2971/3345 [00:37<00:07, 48.24it/s]

decoding POST:  89%|████████▉ | 2982/3345 [00:37<00:06, 57.88it/s]

decoding POST:  89%|████████▉ | 2993/3345 [00:37<00:05, 69.52it/s]

decoding POST:  90%|████████▉ | 3001/3345 [00:37<00:05, 62.33it/s]

decoding POST:  90%|████████▉ | 3008/3345 [00:37<00:05, 60.99it/s]

decoding POST:  90%|█████████ | 3015/3345 [00:37<00:05, 55.71it/s]

decoding POST:  90%|█████████ | 3021/3345 [00:37<00:06, 50.93it/s]

decoding POST:  91%|█████████ | 3029/3345 [00:38<00:05, 54.59it/s]

decoding POST:  91%|█████████ | 3043/3345 [00:38<00:04, 73.47it/s]

decoding POST:  91%|█████████▏| 3054/3345 [00:38<00:03, 76.02it/s]

decoding POST:  92%|█████████▏| 3068/3345 [00:38<00:03, 90.63it/s]

decoding POST:  92%|█████████▏| 3078/3345 [00:38<00:03, 86.59it/s]

decoding POST:  92%|█████████▏| 3090/3345 [00:38<00:02, 87.85it/s]

decoding POST:  93%|█████████▎| 3100/3345 [00:38<00:03, 78.84it/s]

decoding POST:  93%|█████████▎| 3109/3345 [00:38<00:03, 69.71it/s]

decoding POST:  93%|█████████▎| 3117/3345 [00:39<00:03, 68.63it/s]

decoding POST:  93%|█████████▎| 3125/3345 [00:39<00:03, 59.25it/s]

decoding POST:  94%|█████████▎| 3132/3345 [00:39<00:03, 55.07it/s]

decoding POST:  94%|█████████▍| 3141/3345 [00:39<00:03, 59.21it/s]

decoding POST:  94%|█████████▍| 3150/3345 [00:39<00:03, 62.63it/s]

decoding POST:  94%|█████████▍| 3160/3345 [00:39<00:02, 71.06it/s]

decoding POST:  95%|█████████▍| 3170/3345 [00:39<00:02, 72.67it/s]

decoding POST:  95%|█████████▌| 3181/3345 [00:40<00:02, 81.54it/s]

decoding POST:  95%|█████████▌| 3194/3345 [00:40<00:01, 93.64it/s]

decoding POST:  96%|█████████▌| 3207/3345 [00:40<00:01, 101.85it/s]

decoding POST:  96%|█████████▌| 3218/3345 [00:40<00:01, 88.12it/s] 

decoding POST:  97%|█████████▋| 3228/3345 [00:40<00:01, 68.29it/s]

decoding POST:  97%|█████████▋| 3236/3345 [00:40<00:01, 66.87it/s]

decoding POST:  97%|█████████▋| 3244/3345 [00:40<00:01, 66.06it/s]

decoding POST:  97%|█████████▋| 3252/3345 [00:41<00:01, 65.49it/s]

decoding POST:  97%|█████████▋| 3259/3345 [00:41<00:01, 54.86it/s]

decoding POST:  98%|█████████▊| 3270/3345 [00:41<00:01, 62.93it/s]

decoding POST:  98%|█████████▊| 3277/3345 [00:41<00:01, 61.32it/s]

decoding POST:  98%|█████████▊| 3289/3345 [00:41<00:00, 74.73it/s]

decoding POST:  99%|█████████▉| 3311/3345 [00:41<00:00, 103.68it/s]

decoding POST:  99%|█████████▉| 3322/3345 [00:41<00:00, 84.76it/s] 

decoding POST: 100%|█████████▉| 3332/3345 [00:42<00:00, 81.48it/s]

decoding POST: 100%|█████████▉| 3341/3345 [00:42<00:00, 78.74it/s]

decoding POST: 100%|██████████| 3345/3345 [00:42<00:00, 79.21it/s]

### Control: shuffle which cell owns which place field

The strongest test of whether the sequences are real is to destroy the correspondence
between cells and fields while keeping everything else, including the burst times, the
spike counts and the decoding machinery, identical. If the pipeline manufactures
sequences, this control will still find them.

In [14]:
perm_ids = RNG.permutation(len(ids))
shuf = {k: pl.make_template(tcs[k][:, is_place][:, perm_ids], ids, centers) for k in tcs}
ctrl_rows = []
for j in RNG.choice(len(df), size=min(400, len(df)), replace=False):
    row = df.iloc[j]
    out = pl.score_event(spk, shuf, centers,
                         nap.IntervalSet(row.t0, row.t0 + row.dur), RNG)
    if out is not None:
        ctrl_rows.append(out[0])
ctrl = pd.DataFrame(ctrl_rows)
ctrl["significant"] = ctrl["max_p"] < pl.ALPHA
ctrl.to_csv("replay_control.csv", index=False)

In [15]:
tab = [[int(df[(df.epoch == e) & df.significant].shape[0]),
        int(df[(df.epoch == e) & ~df.significant].shape[0])] for e in ("POST", "PRE")]
odds, p_fisher = stats.fisher_exact(tab)
for e in ("PRE", "POST"):
    g = df[df.epoch == e]
    k, n = int(g.significant.sum()), len(g)
    print("%s: %d/%d significant (%.1f%%); vs 5%% chance p=%.2g"
          % (e, k, n, 100 * k / n, stats.binomtest(k, n, 0.05, alternative="greater").pvalue))
print("cell-identity shuffle: %d/%d (%.1f%%)"
      % (ctrl.significant.sum(), len(ctrl), 100 * ctrl.significant.mean()))
print("POST vs PRE: odds ratio %.2f, Fisher p=%.2g" % (odds, p_fisher))

# The cell-ID shuffle, not the nominal 5%, is the operative null here: it lands above
# alpha, which means the per-event test is somewhat liberal (the two posterior shuffles do
# not destroy every source of apparent sequence structure). POST therefore has to beat the
# shuffle rate, not 0.05, before it counts for anything.
n_post_sig = int(df[(df.epoch == "POST")].significant.sum())
n_post = int((df.epoch == "POST").sum())
odds_cs, p_ctrl = stats.fisher_exact([[n_post_sig, n_post - n_post_sig],
                                      [int(ctrl.significant.sum()),
                                       int((~ctrl.significant).sum())]])
print("POST vs cell-ID shuffle: %.1f%% vs %.1f%%, odds ratio %.2f, Fisher p=%.2g"
      % (100 * n_post_sig / n_post, 100 * ctrl.significant.mean(), odds_cs, p_ctrl))
print("(the shuffle sits above the nominal 5%, so it, not alpha, is the baseline to beat)")
sig = df[df.significant]
print("significant events: %d forward, %d reverse" % (sig.fwd.sum(), (~sig.fwd).sum()))
print("median replay speed %.0f cm/s = %.0fx the animal's running speed"
      % (sig.slope.abs().median(), sig.slope.abs().median() / run_speed))

PRE: 62/1023 significant (6.1%); vs 5% chance p=0.072
POST: 147/971 significant (15.1%); vs 5% chance p=2.3e-32
cell-identity shuffle: 35/400 (8.8%)
POST vs PRE: odds ratio 2.77, Fisher p=2.7e-11
POST vs cell-ID shuffle: 15.1% vs 8.8%, odds ratio 1.86, Fisher p=0.0015
(the shuffle sits above the nominal 5%, so it, not alpha, is the baseline to beat)
significant events: 104 forward, 105 reverse
median replay speed 283 cm/s = 7x the animal's running speed


### What a replay event looks like

Three forward and three reverse events from post-task sleep. Top: the raw LFP with the
ripple-filtered trace beneath it. Middle: the posterior over track position, which
sweeps smoothly across the whole 1.6 m in 200-300 ms. Bottom: the same event as a raw
raster with cells ordered by the position of their place field, where the sequence is
visible without any decoding at all.

In [16]:
field_peak = {k: centers[np.argmax(tcs[k][:, is_place], axis=0)] for k in tcs}
strength = sig.assign(abs_r=sig.wcorr.abs())
best6 = pd.concat([
    strength[(strength.epoch == "POST") & strength.fwd].nlargest(3, "abs_r"),
    strength[(strength.epoch == "POST") & ~strength.fwd].nlargest(3, "abs_r"),
])

fig, axes = plt.subplots(3, 6, figsize=(18, 9),
                         gridspec_kw={"height_ratios": [0.55, 1.5, 1.1], "hspace": 0.5, "wspace": 0.38})
fig.subplots_adjust(top=0.80)
for col, (_, ev) in enumerate(best6.iterrows()):
    post, tt = posteriors[(ev.epoch, int(ev.event))]
    t_rel = (tt - tt[0]) * 1000
    pad = 0.05
    sl = slice(int((ev.t0 - pad - t0) * fs), int((ev.t0 + ev.dur + pad - t0) * fs))
    lt = (np.arange(sl.start, sl.stop) / fs + t0 - ev.t0) * 1000

    ax = axes[0, col]
    ax.plot(lt, raw[sl] - raw[sl].mean(), color="k", lw=0.6)
    ax.plot(lt, filt[sl] * 2 - 1500, color="C3", lw=0.6)
    ax.set(xticks=[], yticks=[], xlim=(lt[0], lt[-1]))
    ax.set_title("%s, %sward template\n|r| = %.2f,  p = %.3f\n%.0f cm/s"
                 % ("FORWARD" if ev.fwd else "REVERSE", ev.template,
                    abs(ev.wcorr), ev.max_p, abs(ev.slope)),
                 fontsize=9, pad=8,
                 color="C0" if ev.fwd else "C1")

    ax = axes[1, col]
    ax.imshow(post, aspect="auto", origin="lower", cmap="magma",
              extent=[t_rel[0] - 10, t_rel[-1] + 10, 0, track_len])
    com = (post / post.sum(0) * centers[:, None]).sum(0)
    ax.plot(t_rel, com, "o", color="w", ms=3, alpha=0.8)
    ax.plot(t_rel, np.polyval(np.polyfit(t_rel, com, 1), t_rel), color="C0", lw=1.6)
    ax.set(xlim=(lt[0], lt[-1]), ylim=(0, track_len))
    if col == 0:
        ax.set_ylabel("decoded position (cm)")

    ax = axes[2, col]
    sub_spk = spk.restrict(nap.IntervalSet(ev.t0 - pad, ev.t0 + ev.dur + pad))
    for r_i, u in enumerate(ids[np.argsort(field_peak[ev.template])]):
        s = (np.asarray(sub_spk[u].index) - ev.t0) * 1000
        if len(s):
            ax.plot(s, np.full_like(s, r_i), "|", color="k", ms=3, mew=0.8)
    ax.axvspan(0, ev.dur * 1000, color="C3", alpha=0.10)
    ax.set(xlim=(lt[0], lt[-1]), ylim=(-1, len(ids)), xlabel="time (ms)")
    if col == 0:
        ax.set_ylabel("cell (sorted by field position)")

fig.suptitle("Replay of the linear track inside sharp-wave ripples (post-task sleep)\n"
             "top: raw + ripple-filtered LFP;  middle: posterior P(position | spikes);  bottom: place-cell raster",
             fontsize=11)
fig.savefig("fig06_replay_examples.png", dpi=150, bbox_inches="tight")
plt.close(fig)

### A second, decoding-free line of evidence

Bayesian decoding involves many choices. An older and much simpler measure asks only
whether pairs of cells that fired together on the track also fire together in sleep
ripples: the explained variance EV is the correlation between run and post-sleep
pair-correlation vectors, partialled for pre-sleep, and REV is the same quantity with
pre and post exchanged. EV >> REV is reactivation that cannot be attributed to
pre-existing coupling between cells.

In [17]:
ev_val, rev_val, n_pairs = pl.explained_variance(spk, ripples, run, epochs)
print("EV = %.3f, REV = %.3f over %d cell pairs" % (ev_val, rev_val, n_pairs))

EV = 0.088, REV = 0.001 over 5565 cell pairs


In [18]:
fig = plt.figure(figsize=(14, 8.5))
gs = fig.add_gridspec(2, 3, hspace=0.42, wspace=0.32)

ax = fig.add_subplot(gs[0, 0])
ax.hist(err, bins=np.arange(0, track_len, 4), color="C0")
ax.axvline(np.median(err), color="k", ls="--")
ax.axvline(chance, color="C3", ls=":")
ax.text(np.median(err) + 4, ax.get_ylim()[1] * 0.9, "median %.1f cm" % np.median(err), fontsize=9)
ax.text(chance + 4, ax.get_ylim()[1] * 0.6, "chance", color="C3", fontsize=9)
ax.set(xlabel="decoding error while running (cm)", ylabel="count",
       title="The template decodes real position\n(250 ms bins, MAZE epoch)")

ax = fig.add_subplot(gs[0, 1])
labels = ["PRE sleep", "POST sleep", "cell-ID\nshuffle"]
vals = [100 * df[df.epoch == "PRE"].significant.mean(),
        100 * df[df.epoch == "POST"].significant.mean(),
        100 * ctrl.significant.mean()]
ns = [int((df.epoch == "PRE").sum()), int((df.epoch == "POST").sum()), len(ctrl)]
bars = ax.bar(labels, vals, color=["0.6", "C0", "0.85"])
ax.axhline(5, color="C3", ls="--", lw=1, label="nominal α = 0.05")
ax.axhline(vals[2], color="k", ls="-.", lw=1.2,
           label="empirical null (cell-ID shuffle, %.1f%%)" % vals[2])
for b, v, n in zip(bars, vals, ns):
    ax.text(b.get_x() + b.get_width() / 2, v + 0.4, "%.1f%%\n(n=%d)" % (v, n), ha="center", fontsize=8)
ax.legend(fontsize=7.5, frameon=False, loc="upper left")
ax.set(ylabel="% of ripple events with significant replay", ylim=(0, max(vals) * 1.55),
       title="POST beats PRE and the cell-ID shuffle\nvs PRE p=%.1g,  vs shuffle p=%.1g"
             % (p_fisher, p_ctrl))

ax = fig.add_subplot(gs[0, 2])
bins = np.linspace(0, 1, 26)
for name, v, c in (("POST", df[df.epoch == "POST"].wcorr.abs(), "C0"),
                   ("PRE", df[df.epoch == "PRE"].wcorr.abs(), "0.5"),
                   ("cell-ID shuffle", ctrl.wcorr.abs(), "C3")):
    ax.hist(v, bins=bins, histtype="step", density=True, lw=1.8, color=c, label=name)
ax.legend(fontsize=8, frameon=False)
ax.set(xlabel="|weighted correlation|", ylabel="density", title="Sequence scores by epoch")

ax = fig.add_subplot(gs[1, 0])
ax.hist(sig.slope.abs() / 100, bins=np.arange(0, 9, 0.4), color="C0")
ax.axvline(run_speed / 100, color="C3", ls="--")
ax.set(xlabel="replay speed (m/s)", ylabel="count", xlim=(0, 9),
       title="Replay is time-compressed\nmedian %.1f m/s = %.0fx running"
             % (sig.slope.abs().median() / 100, sig.slope.abs().median() / run_speed))
ax.text(run_speed / 100 + 0.25, ax.get_ylim()[1] * 0.92,
        "running speed\n%.1f m/s" % (run_speed / 100), color="C3", fontsize=8, va="top")

ax = fig.add_subplot(gs[1, 1])
x = np.arange(2)
ax.bar(x - 0.18, [int(((sig.epoch == e) & sig.fwd).sum()) for e in ("PRE", "POST")], 0.36,
       label="forward", color="C0")
ax.bar(x + 0.18, [int(((sig.epoch == e) & ~sig.fwd).sum()) for e in ("PRE", "POST")], 0.36,
       label="reverse", color="C1")
ax.set_xticks(x)
ax.set_xticklabels(["PRE", "POST"])
ax.legend(fontsize=8, frameon=False)
ax.set(ylabel="significant events",
       title="Forward and reverse replay\nboth occur, in comparable numbers")

ax = fig.add_subplot(gs[1, 2])
ax.bar(["EV\n(POST | RUN, PRE)", "REV\n(PRE | RUN, POST)"], [ev_val, rev_val], color=["C0", "0.6"])
for i, v in enumerate([ev_val, rev_val]):
    ax.text(i, v + 0.002, "%.3f" % v, ha="center", fontsize=9)
ax.set(ylabel="explained variance", title="Pairwise reactivation, no decoding\n%d cell pairs" % n_pairs)

fig.suptitle("Hippocampal replay during sharp-wave ripples, %s (DANDI:000044)" % SESSION)
fig.savefig("fig07_replay_summary.png", dpi=150, bbox_inches="tight")
plt.close(fig)

## 6. Four sessions, three rats

The whole pipeline (channel selection, ripple detection, place fields, decoding,
shuffles, controls) is repeated on three further sessions. `Achilles-11012013` is
excluded because its maze is circular rather than linear, which would need a different
linearization. Results are cached to CSV so re-running the notebook is cheap.

This is where the two halves of the result come apart, and it is worth stating plainly
rather than burying. The ripple physiology reproduces almost exactly across all four
sessions. The replay statistic does not: only `Achilles-10252013`, which has two to
three times the place-cell yield of the others, shows a POST fraction clearly above both
its PRE fraction and its own cell-ID shuffle. In the other three sessions the POST, PRE
and shuffle fractions all sit within a couple of percentage points of each other, and
the cell-ID shuffle itself runs at 6-7.5% rather than the nominal 5%, which means the
per-event test is slightly liberal and small apparent enrichments there should not be
read as replay.

In [19]:
SESSIONS = ["Achilles-10252013", "Gatsby-08022013", "Cicero-09172014", "Buddy-06272013"]

if os.path.exists("multi_session_summary.csv"):
    res = pd.read_csv("multi_session_summary.csv")
    all_events = pd.read_csv("multi_session_events.csv")
else:
    rows, frames = [], []
    for s in SESSIONS:
        summary, d, c, _ = pl.run_session(s)
        summary["frac_ctrl_n"] = len(c)
        rows.append(summary)
        d["session"] = s
        frames.append(d)
    res = pd.DataFrame(rows)
    all_events = pd.concat(frames)
    res.to_csv("multi_session_summary.csv", index=False)
    all_events.to_csv("multi_session_events.csv", index=False)

print(res[["session", "n_place", "n_ripples", "rate_nrem", "rate_rem", "decoder_err_cm",
           "frac_pre", "frac_post", "frac_ctrl", "ev", "rev"]].to_string(index=False))

# Per-session POST-vs-PRE tests, plus a Cochran-Mantel-Haenszel test stratified by
# session. Naively pooling events across sessions would be misleading here, because the
# sessions differ enormously in place-cell yield; CMH keeps sessions as strata and its
# companion homogeneity test says explicitly whether one effect size fits all of them.
import statsmodels.api as sm

tabs, per_session = [], []
for s in res.session:
    g = all_events[all_events.session == s]
    t = [[int(((g.epoch == e) & g.significant).sum()),
          int(((g.epoch == e) & ~g.significant).sum())] for e in ("POST", "PRE")]
    orr, pv = stats.fisher_exact(t)
    per_session.append(dict(session=s, n_place=int(res.loc[res.session == s, "n_place"].iloc[0]),
                            post_sig=t[0][0], post_n=t[0][0] + t[0][1],
                            pre_sig=t[1][0], pre_n=t[1][0] + t[1][1],
                            odds_ratio=orr, p=pv))
    tabs.append(np.array(t).T)
per_session = pd.DataFrame(per_session)
print("\nPOST vs PRE, session by session:")
print(per_session.to_string(index=False,
                            formatters={"odds_ratio": "{:.2f}".format, "p": "{:.3g}".format}))

cmh = sm.stats.StratifiedTable(np.dstack(tabs))
or_cmh = cmh.oddsratio_pooled
p_cmh = cmh.test_null_odds().pvalue
p_homog = cmh.test_equal_odds().pvalue
print("\nCochran-Mantel-Haenszel (stratified by session): OR=%.2f, p=%.2g" % (or_cmh, p_cmh))
print("homogeneity of the odds ratios across sessions: p=%.3g" % p_homog)
print("-> the effect is NOT homogeneous; %s carries it."
      % per_session.loc[per_session.odds_ratio.idxmax(), "session"])

          session  n_place  n_ripples  rate_nrem  rate_rem  decoder_err_cm  frac_pre  frac_post  frac_ctrl       ev      rev
Achilles-10252013      106      14625   0.612649  0.001449        4.729924  0.059629   0.151390     0.0625 0.088179 0.000794
  Gatsby-08022013       44      14838   0.648649  0.009881        5.457835  0.078459   0.083532     0.0750 0.057690 0.000952
  Cicero-09172014       30      12402   0.691570  0.016611        8.334354  0.061181   0.096774     0.0725 0.023842 0.005439
   Buddy-06272013       35      10731   0.575518  0.038366        7.818839  0.049793   0.063898     0.0600 0.035173 0.003050



POST vs PRE, session by session:
          session  n_place  post_sig  post_n  pre_sig  pre_n odds_ratio        p
Achilles-10252013      106       147     971       61   1023       2.81 1.43e-11
  Gatsby-08022013       44        35     419       55    701       1.07     0.82
  Cicero-09172014       30        45     465       29    474       1.64   0.0521
   Buddy-06272013       35        20     313       24    482       1.30    0.429

Cochran-Mantel-Haenszel (stratified by session): OR=1.88, p=1.7e-09
homogeneity of the odds ratios across sessions: p=0.0023
-> the effect is NOT homogeneous; Achilles-10252013 carries it.


In [20]:
labels = [s.split("-")[0] for s in res.session]
x = np.arange(len(res))
fig, axes = plt.subplots(2, 3, figsize=(14, 7.5))
fig.subplots_adjust(hspace=0.5, wspace=0.3)

ax = axes[0, 0]
for i, (k, c) in enumerate((("rate_nrem", "C0"), ("rate_awake", "C2"), ("rate_rem", "C1"))):
    ax.bar(x + (i - 1) * 0.27, res[k], 0.27, label=k.replace("rate_", ""), color=c)
ax.set_xticks(x, labels, rotation=15)
ax.legend(fontsize=8, frameon=False)
ax.set(ylabel="ripple rate (Hz)", title="Ripple rate by brain state")

ax = axes[0, 1]
ax.bar(x - 0.2, res.ripple_dur_ms, 0.4, color="C0", label="duration (ms)")
ax.bar(x + 0.2, res.ripple_freq_hz, 0.4, color="C3", label="frequency (Hz)")
ax.set_xticks(x, labels, rotation=15)
ax.legend(fontsize=8, frameon=False)
ax.set(ylabel="ms  /  Hz", title="Ripple duration and frequency")

ax = axes[0, 2]
ax.bar(x - 0.2, res.n_place, 0.4, label="place cells", color="C0")
ax.bar(x + 0.2, res.decoder_err_cm, 0.4, label="decoder error (cm)", color="C1")
ax.set_xticks(x, labels, rotation=15)
ax.legend(fontsize=8, frameon=False)
ax.set(ylabel="count  /  cm", title="Template quality")

ax = axes[1, 0]
ax.bar(x - 0.27, 100 * res.frac_pre, 0.27, label="PRE sleep", color="0.6")
ax.bar(x, 100 * res.frac_post, 0.27, label="POST sleep", color="C0")
ax.bar(x + 0.27, 100 * res.frac_ctrl, 0.27, label="cell-ID shuffle", color="0.85")
ax.axhline(5, color="C3", ls="--", lw=1, label="nominal α = 0.05")
for i, row in per_session.reset_index(drop=True).iterrows():
    star = "p<1e-10" if row.p < 1e-10 else ("p=%.2f" % row.p)
    ax.text(i, 100 * res.frac_post.iloc[i] + 0.4, star, ha="center", fontsize=7.5,
            fontweight="bold" if row.p < 0.05 else "normal",
            color="k" if row.p < 0.05 else "0.45")
ax.set_xticks(x, labels, rotation=15)
ax.legend(fontsize=7.5, frameon=False, loc="upper right")
ax.set_ylim(0, 100 * max(res.frac_post.max(), res.frac_pre.max()) * 1.35)
ax.set(ylabel="% events with significant replay",
       title="POST > PRE only in Achilles\n(CMH OR=%.2f, but homogeneity p=%.3f)"
             % (or_cmh, p_homog))

ax = axes[1, 1]
ax.bar(x - 0.2, res.ev, 0.4, label="EV", color="C0")
ax.bar(x + 0.2, res.rev, 0.4, label="REV", color="0.6")
ax.set_xticks(x, labels, rotation=15)
ax.legend(fontsize=8, frameon=False)
ax.set(ylabel="explained variance", title="Pairwise reactivation (EV > REV)")

ax = axes[1, 2]
for s in res.session:
    g = all_events[(all_events.session == s) & all_events.significant]
    ax.hist(np.abs(g.slope) / 100, bins=np.arange(0, 12, 0.75), histtype="step", lw=1.6,
            density=True, label=s.split("-")[0])
ax.legend(fontsize=8, frameon=False)
ax.set(xlabel="replay speed (m/s)", ylabel="density", title="Replay speed")

fig.suptitle("Sharp-wave ripples and replay across four sessions of DANDI:000044")
fig.savefig("fig08_multi_session.png", dpi=150, bbox_inches="tight")
plt.close(fig)

## 7. What the data show

The LFP events isolated here are sharp-wave ripples by every standard criterion: a
~50 ms, ~160 Hz oscillation superimposed on a slow sharp wave, occurring at roughly
0.5 Hz in non-REM sleep, at a lower rate during quiet waking, and essentially never in
REM; accompanied by a large increase in pyramidal and (much larger) interneuron firing;
and detected identically from an electrode on a different shank.

The spikes inside those ripples carry the track. Decoding population bursts against the
running place-field template recovers smooth trajectories across the maze at roughly
2-3 m/s, some replaying the run in the direction it was experienced and some in
reverse. In `Achilles-10252013` the fraction of ripple events carrying a statistically
significant trajectory is about 15% in the sleep that follows the run, against about 6%
in the sleep that precedes it (odds ratio 2.8, p = 3e-11) and about 9% in a control that
permutes cell identities across place fields (p = 0.002). The same asymmetry appears in a
completely different statistic that involves no decoding: pairwise co-firing during
ripples resembles co-firing on the track far more in post-task sleep than in pre-task
sleep (EV = 0.088, REV = 0.001).

The cell-identity control is the number worth dwelling on. At 9% it sits well above the
nominal 5%, which means the per-event test is liberal and that roughly half of the
"significant" POST events would be expected from a pipeline that had no true
cell-to-field correspondence at all. The POST excess over that control is still clear in
this session, but it is an excess over 9%, not over 5%, and the exact counts move by a
few events between runs because the shuffle uses only 400 draws.

The honest reading of the four-session comparison is that these two results have very
different evidential strength. The ripple physiology is reproducible: rate, duration,
intra-ripple frequency and the non-REM / REM contrast are nearly identical in all four
sessions. The replay result is carried by one session. Achilles is individually
significant; Cicero is marginal (OR 1.6, p = 0.05); Gatsby (OR 1.07) and Buddy (OR 1.3)
are not. The stratified CMH odds ratio of 1.9 is significant, but its homogeneity test
rejects at p = 0.002, which is the formal way of saying that a single pooled number does
not describe these four sessions and that quoting it alone would overstate the case.

The most likely reason is simply the size of the ensemble. Achilles contributes 106
place cells and decodes the animal's real position to 4.7 cm; the other sessions
contribute 30 to 44 place cells and decode to 5.5-8.3 cm. Weighted correlation over a
20 ms bin needs enough simultaneously active fields to distinguish a trajectory from a
scattered posterior, so a sequence measure that works at 106 cells can fall to near
chance at 30 without the underlying phenomenon changing at all. That is an interpretable
limitation of the measurement rather than evidence against replay, but it is a
limitation, and the pairwise EV/REV statistic (which needs far fewer cells and does show
EV > REV in all four sessions) is the better-supported cross-session claim here.

In [21]:
with open("results_summary.txt", "w") as f:
    f.write("session: %s\n" % SESSION)
    f.write("ripples: %d (%.2f Hz non-REM, %.3f Hz REM)\n" % (len(peak_t), rates["Non-REM"], rates["REM"]))
    f.write("ripple duration %.0f ms, frequency %.0f Hz\n" % (np.median(dur_ms), np.median(peak_freq)))
    f.write("place cells: %d of %d pyramidal\n" % (is_place.sum(), len(pyr)))
    f.write("decoder median error %.1f cm (chance %.1f cm)\n" % (np.median(err), chance))
    for e in ("PRE", "POST"):
        g = df[df.epoch == e]
        f.write("%s: %d/%d significant replay (%.1f%%)\n"
                % (e, g.significant.sum(), len(g), 100 * g.significant.mean()))
    f.write("cell-ID shuffle: %.1f%% (n=%d)  <- empirical null, above nominal 5%%\n"
            % (100 * ctrl.significant.mean(), len(ctrl)))
    f.write("Fisher POST vs PRE: OR=%.2f p=%.3g\n" % (odds, p_fisher))
    f.write("Fisher POST vs cell-ID shuffle: OR=%.2f p=%.3g\n" % (odds_cs, p_ctrl))
    f.write("replay speed %.0f cm/s vs running %.0f cm/s\n" % (sig.slope.abs().median(), run_speed))
    f.write("EV=%.3f REV=%.3f (%d pairs)\n" % (ev_val, rev_val, n_pairs))
    f.write("\n4 sessions, POST vs PRE:\n")
    for _, r in per_session.iterrows():
        f.write("  %-18s %d/%d POST vs %d/%d PRE  OR=%.2f p=%.3g\n"
                % (r.session, r.post_sig, r.post_n, r.pre_sig, r.pre_n, r.odds_ratio, r.p))
    f.write("CMH OR=%.2f p=%.3g; homogeneity p=%.3g (effect is NOT homogeneous)\n"
            % (or_cmh, p_cmh, p_homog))
print(open("results_summary.txt").read())

session: Achilles-10252013
ripples: 14625 (0.61 Hz non-REM, 0.001 Hz REM)
ripple duration 42 ms, frequency 162 Hz
place cells: 106 of 120 pyramidal
decoder median error 4.7 cm (chance 46.8 cm)
PRE: 62/1023 significant replay (6.1%)
POST: 147/971 significant replay (15.1%)
cell-ID shuffle: 8.8% (n=400)  <- empirical null, above nominal 5%
Fisher POST vs PRE: OR=2.77 p=2.66e-11
Fisher POST vs cell-ID shuffle: OR=1.86 p=0.00155
replay speed 283 cm/s vs running 42 cm/s
EV=0.088 REV=0.001 (5565 pairs)

4 sessions, POST vs PRE:
  Achilles-10252013  147/971 POST vs 61/1023 PRE  OR=2.81 p=1.43e-11
  Gatsby-08022013    35/419 POST vs 55/701 PRE  OR=1.07 p=0.82
  Cicero-09172014    45/465 POST vs 29/474 PRE  OR=1.64 p=0.0521
  Buddy-06272013     20/313 POST vs 24/482 PRE  OR=1.30 p=0.429
CMH OR=1.88 p=1.71e-09; homogeneity p=0.0023 (effect is NOT homogeneous)

